// wollok-ts/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WOLLOK_EXTRA_STACK_TRACE_HEADER = exports.REPL = exports.KEYWORDS = exports.INFIX_OPERATORS = exports.ASSIGNATION_OPERATORS = exports.PREFIX_OPERATORS = exports.CLOSURE_TO_STRING_METHOD = exports.CLOSURE_EVALUATE_METHOD = exports.APPLY_METHOD = exports.TO_STRING_METHOD = exports.INITIALIZE_METHOD = exports.GAME_MODULE = exports.VOID_WKO = exports.CLOSURE_MODULE = exports.EXCEPTION_MODULE = exports.OBJECT_MODULE = exports.DICTIONARY_MODULE = exports.RANGE_MODULE = exports.PAIR_MODULE = exports.DATE_MODULE = exports.STRING_MODULE = exports.NUMBER_MODULE = exports.BOOLEAN_MODULE = exports.SET_MODULE = exports.LIST_MODULE = exports.COLLECTION_MODULE = exports.WOLLOK_FILE_EXTENSION = exports.PROGRAM_FILE_EXTENSION = exports.TEST_FILE_EXTENSION = exports.WOLLOK_BASE_PACKAGE = void 0;
exports.WOLLOK_BASE_PACKAGE = 'wollok.';
exports.TEST_FILE_EXTENSION = 'wtest';
exports.PROGRAM_FILE_EXTENSION = 'wpgm';
exports.WOLLOK_FILE_EXTENSION = 'wlk';
exports.COLLECTION_MODULE = 'wollok.lang.Collection';
exports.LIST_MODULE = 'wollok.lang.List';
exports.SET_MODULE = 'wollok.lang.Set';
exports.BOOLEAN_MODULE = 'wollok.lang.Boolean';
exports.NUMBER_MODULE = 'wollok.lang.Number';
exports.STRING_MODULE = 'wollok.lang.String';
exports.DATE_MODULE = 'wollok.lang.Date';
exports.PAIR_MODULE = 'wollok.lang.Pair';
exports.RANGE_MODULE = 'wollok.lang.Range';
exports.DICTIONARY_MODULE = 'wollok.lang.Dictionary';
exports.OBJECT_MODULE = 'wollok.lang.Object';
exports.EXCEPTION_MODULE = 'wollok.lang.Exception';
exports.CLOSURE_MODULE = 'wollok.lang.Closure';
exports.VOID_WKO = 'wollok.lang.void';
exports.GAME_MODULE = 'wollok.game.game';
exports.INITIALIZE_METHOD = 'initialize';
exports.TO_STRING_METHOD = 'toString';
exports.APPLY_METHOD = 'apply';
exports.CLOSURE_EVALUATE_METHOD = '<apply>';
exports.CLOSURE_TO_STRING_METHOD = '<toString>';
exports.PREFIX_OPERATORS = {
    '!': 'negate',
    '-': 'invert',
    '+': 'plus',
    'not': 'negate',
};
exports.ASSIGNATION_OPERATORS = ['=', '||=', '/=', '-=', '+=', '*=', '&&=', '%='];
exports.INFIX_OPERATORS = [
    ['||', 'or'],
    ['&&', 'and'],
    ['===', '==', '!==', '!='],
    ['>=', '>', '<=', '<'],
    ['?:', '>>>', '>>', '>..', '<>', '<=>', '<<<', '<<', '..<', '..', '->'],
    ['-', '+'],
    ['/', '*'],
    ['**', '%'],
];
exports.KEYWORDS = {
    IF: 'if',
    ELSE: 'else',
    NEW: 'new',
    SELF: 'self',
    SUPER: 'super',
    NULL: 'null',
    METHOD: 'method',
    VAR: 'var',
    THROW: 'throw',
    TRY: 'try',
    CATCH: 'catch',
    THEN: 'then',
    ALWAYS: 'always',
    PROPERTY: 'property',
    CONST: 'const',
    OVERRIDE: 'override',
    NATIVE: 'native',
    RETURN: 'return',
    IMPORT: 'import',
    SUITE: 'describe',
    TEST: 'test',
    MIXIN: 'mixin',
    CLASS: 'class',
    INHERITS: 'inherits',
    DERIVED: 'derived',
    MIXED_AND: 'and',
    WKO: 'object',
    FIXTURE: 'fixture',
    PROGRAM: 'program',
    PACKAGE: 'package',
    ONLY: 'only',
};
exports.REPL = 'REPL';
exports.WOLLOK_EXTRA_STACK_TRACE_HEADER = 'Derived from TypeScript stack';
