// wollok-ts/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.natives = exports.WRENatives = exports.WRE = exports.print = exports.validate = exports.ParseError = exports.parse = exports.link = exports.buildEnvironment = void 0;
const linker_1 = __importDefault(require("./linker"));
exports.link = linker_1.default;
const jsonUtils_1 = require("./jsonUtils");
const parse = __importStar(require("./parser"));
exports.parse = parse;
const parser_1 = require("./parser");
Object.defineProperty(exports, "ParseError", { enumerable: true, get: function () { return parser_1.ParseError; } });
const validator_1 = __importDefault(require("./validator"));
exports.validate = validator_1.default;
const print_1 = __importDefault(require("./printer/print"));
exports.print = print_1.default;
const wre_json_1 = __importDefault(require("./wre/wre.json"));
exports.WRE = wre_json_1.default;
const wre_natives_1 = __importDefault(require("./wre/wre.natives"));
exports.WRENatives = wre_natives_1.default;
const natives_1 = __importDefault(require("./wre/natives"));
exports.natives = natives_1.default;
function buildEnvironment(files, baseEnvironment = (0, jsonUtils_1.fromJSON)(wre_json_1.default)) {
    return (0, linker_1.default)(files.map(({ name, content }) => {
        try {
            return parse.File(name).tryParse(content);
        }
        catch (error) {
            throw new Error(`Failed to parse ${name}: ${error.message}`);
        }
    }), baseEnvironment);
}
exports.buildEnvironment = buildEnvironment;
__exportStar(require("./constants"), exports);
__exportStar(require("./extensions"), exports);
__exportStar(require("./helpers"), exports);
__exportStar(require("./linker"), exports);
__exportStar(require("./jsonUtils"), exports);
__exportStar(require("./model"), exports);
__exportStar(require("./interpreter/interpreter"), exports);
__exportStar(require("./interpreter/dynamicDiagram"), exports);
__exportStar(require("./interpreter/runtimeModel"), exports);
__exportStar(require("./typeSystem/constraintBasedTypeSystem"), exports);
__exportStar(require("./printer/exceptions"), exports);
__exportStar(require("./printer/utils"), exports);
__exportStar(require("./validator/messageReporter"), exports);
