<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Manual &mdash; EggLib</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.1.11',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/icon.ico"/>
    <link rel="top" title="EggLib" href="index.html" />
    <link rel="next" title="eggcoal" href="eggcoal.html" />
    <link rel="prev" title="Download and install" href="installation.html" /> 
  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px 0px 0px 0px">
<a href="index.html"><img src="_static/egglib_banniere5.png" border="0" alt="EggLib"/></a>
</div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="eggcoal.html" title="eggcoal"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Download and install"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">EggLib</a> &raquo;</li> 
      </ul>
    </div>
<div style="background-color: grey; text-align: left; padding: 0px 0px 0px 10px; border-bottom: 1px solid #2d2e91">
<font color="white" ><b>EggLib version 2.1 is archived.</b></font></div>


      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Manual</a><ul>
<li><a class="reference internal" href="#import-export-of-sequence-data">Import/export of sequence data</a><ul>
<li><a class="reference internal" href="#import-from-fasta-formatted-files-or-strings">Import from FASTA-formatted files or strings</a></li>
<li><a class="reference internal" href="#import-population-information-as-group-label">Import population information as group label</a></li>
<li><a class="reference internal" href="#conversion-to-container-and-align-types">Conversion to Container and Align types</a></li>
<li><a class="reference internal" href="#exports-sequence-data">Exports sequence data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#manipulation-of-sequence-sets-and-alignments">Manipulation of sequence sets and alignments</a><ul>
<li><a class="reference internal" href="#iteration">Iteration</a></li>
<li><a class="reference internal" href="#sequence-data-edition">Sequence data edition</a></li>
</ul>
</li>
<li><a class="reference internal" href="#manipulation-of-other-data-types-trees-annotated-sequences-ssr">Manipulation of other data types (trees, annotated sequences, SSR)</a></li>
<li><a class="reference internal" href="#analysis-of-polymorphism">Analysis of polymorphism</a></li>
<li><a class="reference internal" href="#coalescent-simulations">Coalescent simulations</a></li>
<li><a class="reference internal" href="#interactive-usage-and-approximate-bayesian-computation">Interactive usage and Approximate Bayesian Computation</a><ul>
<li><a class="reference internal" href="#the-egglib-script">The egglib script</a></li>
<li><a class="reference internal" href="#abc-sampling">ABC sampling</a></li>
<li><a class="reference internal" href="#abc-rejection-regression-procedure">ABC rejection-regression procedure</a></li>
<li><a class="reference internal" href="#comparing-abc-models">Comparing ABC models</a></li>
<li><a class="reference internal" href="#analyzing-posterior-distribution">Analyzing posterior distribution</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Download and install</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="eggcoal.html"
                        title="next chapter">eggcoal</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/manual.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="manual">
<span id="id1"></span><h1>Manual<a class="headerlink" href="#manual" title="Permalink to this headline">¶</a></h1>
<p>This section provides primers on the usage of the Python module and of
the executable commands. In the following, representative examples of
the most common uses of Egglib will be addressed, assuming that the user
aims at developing simple scripts (that is, without extensive error
checking).</p>
<div class="section" id="import-export-of-sequence-data">
<h2>Import/export of sequence data<a class="headerlink" href="#import-export-of-sequence-data" title="Permalink to this headline">¶</a></h2>
<p>Egglib relies heavily on two classes: sequence sets (that are not
considered as aligned), <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a>, and sequence
alignments (for which all sequences must have the same lengths),
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>. Many operations are available as methods of these
two classes (including polymorphism analysis).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">There is no sequence validity check at the level of storage
classes. The parser only skips all white spaces and imports
characters as upper cases, but it accepts all characters, and
all characters are allowed when setting sequences within
Python. Checking is performed at latter stages.</p>
</div>
<div class="section" id="import-from-fasta-formatted-files-or-strings">
<h3>Import from FASTA-formatted files or strings<a class="headerlink" href="#import-from-fasta-formatted-files-or-strings" title="Permalink to this headline">¶</a></h3>
<p>Egglib is based primarily on Pearson&#8217;s FASTA format for data input, due
to its simplicity and the fact that it can be generated by many other
software. In Python, the constructor of both <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a>
and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> includes a parser that can take a file name.
Creating a sequence set/alignment objects is then concise. Assuming the
file <tt class="docutils literal"><span class="pre">data1.fas</span></tt> contain a sequence alignment formatted as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;AY446407
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
--------CACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446408
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
--------CACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446409
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
--------CACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446410
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
--------CACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446411
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
--------CACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446412
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
----CACACACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446413
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
----CACACACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446414
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
----CACACACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446415
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACA--
----CACACACACACACGTGGCACTTTCTTCCTGCGGCAC
&gt;AY446416
GAGGCTCATCAGCCTGTTATTTACTGAAATTGAATGAAAAATGAGAGAGACGAGAAATGA
GAAAAA-AATAAAATAAA-----ATAAAATAGTTCAGTTATGGATAAGCAAATACACAC-
-ACACACACACACACACGTGGCACTTTCTTCCTGCGGCAC
</pre></div>
</div>
<p>The alignment can be imported as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">egglib</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="p">(</span><span class="s">&quot;data1.fas&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The constructor of <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> requires that all
sequences have the same length. For non-aligned data (sequence sets),
use the <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> class instead. It is also possible
to create an instance from a fasta-formatted string loaded in memory,
as in the example below (this let you modify the string, if needed):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;data1.fas&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="p">(</span><span class="n">string</span><span class="o">=</span><span class="n">fasta</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="import-population-information-as-group-label">
<h3>Import population information as group label<a class="headerlink" href="#import-population-information-as-group-label" title="Permalink to this headline">¶</a></h3>
<p>Both <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> can import
<em>group labels</em>, which are special tags present in the sequence names.
This feature is disabled by default (because such tags are not defined
in the FASTA format). Note that group labels are always present in
<a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instances and that
they are initialized to the default value <tt class="docutils literal"><span class="pre">0</span></tt> if the option described
hereafter is not activated.</p>
<p>You can activate group label detection by using the flag option <em>groups</em>
of the constructors. In such case, do avoid to use <tt class="docutils literal"><span class="pre">&#64;</span></tt> anywhere in
sequence names. If names in your imported file or string contains labels
such as <tt class="docutils literal"><span class="pre">&#64;0</span></tt>, <tt class="docutils literal"><span class="pre">&#64;1</span></tt>, <tt class="docutils literal"><span class="pre">&#64;42</span></tt> (<tt class="docutils literal"><span class="pre">&#64;</span></tt> followed by an integer, either at
the end of the string or followed by a space), they will be recognized.
Assume that <tt class="docutils literal"><span class="pre">data2.fas</span></tt> contains such labels, as reproduced below
(sequences have been omitted):</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;AY446407
&gt;AY446408@0
&gt;AY446409@1
&gt;AY446410 @18
&gt;AY446411@4 @6
&gt;AY446412@X
&gt;AY446413@7 @X
&gt;AY446414@7 @0
&gt;AY446415@42 POP42
&gt;AY446416@8POP8
</pre></div>
</div>
<p>Some labels have been deliberately placed in strange positions or
repeated. Now you can import the data and check how the names have been
processed (the iterator will be addressed in the next section):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="p">(</span><span class="s">&quot;data2.fas&quot;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">name</span><span class="p">,</span><span class="n">sequence</span><span class="p">,</span><span class="n">group</span> <span class="ow">in</span> <span class="n">align</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">print</span> <span class="s">&#39;name:&#39;</span><span class="p">,</span><span class="n">name</span><span class="p">,</span> <span class="s">&#39;group:&#39;</span><span class="p">,</span><span class="n">group</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre>name: AY446407 group: 0
name: AY446408 group: 0
name: AY446409 group: 1
name: AY446410  group: 18
name: AY446411 group: 46
name: AY446412 group: 0
name: AY446413 group: 7
name: AY446414 group: 1
name: AY446415POP42 group: 42
name: AY446416 group: 8
</pre></div>
</div>
<p>Note first that the names have been modified: the group index tags have
been removed from the name and, in the case of malformed, repeated tags
or tags not placed at the end of the sequence, it will not be possible
to generate the exact same file anymore (the last sequence, for example,
will be <tt class="docutils literal"><span class="pre">AY446416&#64;8</span></tt>. Note also that the group labels are really
labels (not all values need to be used, and the order is not relevant).
Also, importantly, group labels are always integers.</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">AY446407</span></tt>: This has been interpreted as <tt class="docutils literal"><span class="pre">AY446407&#64;0</span></tt> (the
value <tt class="docutils literal"><span class="pre">0</span></tt> is the default).</li>
<li><tt class="docutils literal"><span class="pre">AY446408&#64;0</span></tt>: The group label is <tt class="docutils literal"><span class="pre">0</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">AY446409&#64;1</span></tt>: The group label is <tt class="docutils literal"><span class="pre">1</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">AY446410</span> <span class="pre">&#64;18</span></tt>: The group label is <tt class="docutils literal"><span class="pre">18</span></tt>. Note that many values
have been skipped. It will never be assumed that populations 2
through 17 exist. Note also that there is a space between the
name itself and the tag, and that the space is still present in
the sequence name: the name really is <tt class="docutils literal"><span class="pre">&quot;AY446410</span> <span class="pre">&quot;</span></tt>. <strong>This is
error-prone and should be avoided.</strong></li>
<li><tt class="docutils literal"><span class="pre">AY446411&#64;4</span> <span class="pre">&#64;6</span></tt>: Two labels have been entered, and they have
been concatenated to form the integer <tt class="docutils literal"><span class="pre">46</span></tt>. This syntax is not
recommended.</li>
<li><tt class="docutils literal"><span class="pre">AY446412&#64;X</span></tt>: The label is not integer: the data was lost.</li>
<li><tt class="docutils literal"><span class="pre">AY446413&#64;7</span> <span class="pre">&#64;X</span></tt>: One of the labels is not an integer, this was
lost (avoid using repeated tags).</li>
<li><tt class="docutils literal"><span class="pre">AY446414&#64;7</span> <span class="pre">&#64;0</span></tt>: again a repeated tag. Note that, in the
previous example, <tt class="docutils literal"><span class="pre">&#64;X</span></tt> really was ignored (it was not changed
into <tt class="docutils literal"><span class="pre">0</span></tt>).</li>
<li><tt class="docutils literal"><span class="pre">AY446415&#64;42</span> <span class="pre">POP42</span></tt>: Additional information is added after the
group label. One the tag and the space after it were removed.
This syntax is not recommended.</li>
<li><tt class="docutils literal"><span class="pre">AY446416&#64;8POP8</span></tt>: the space after the group label tag was
omitted. Data has been lost.</li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When using <tt class="docutils literal"><span class="pre">groups=True</span></tt> when importing fasta data, always
place group label tags at the end of the name without space.
It will prevent lots of errors and allow consistent
input/output operations (the input name can be generated again
when exporting), which is the preferable behaviour.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The group label <tt class="docutils literal"><span class="pre">999</span></tt> (entered as <tt class="docutils literal"><span class="pre">&#64;999</span></tt> in input file)
has a special meaning: it identifies an outgroup sequence. It
is possible to enter several outgroup sequences.</p>
</div>
</div>
<div class="section" id="conversion-to-container-and-align-types">
<h3>Conversion to Container and Align types<a class="headerlink" href="#conversion-to-container-and-align-types" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="xref py py-meth docutils literal"><span class="pre">create()</span></tt> method can be called on either
<a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> or <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> classes to create
corresponding instances. <tt class="xref py py-meth docutils literal"><span class="pre">create()</span></tt> expects as argument an object of
any iterable type, provided that the iterations return either
<tt class="docutils literal"><span class="pre">(name,</span> <span class="pre">sequence)</span></tt> or <tt class="docutils literal"><span class="pre">(name,</span> <span class="pre">sequence,</span> <span class="pre">group)</span></tt> tuples. Note that
both <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> are
compatible. Therefore, <tt class="xref py py-meth docutils literal"><span class="pre">create()</span></tt> can be used to perform a deep
copy of sequence sets/alignments or convert one type into another:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">copy</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">container</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Container</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
</pre></div>
</div>
<p>Convert from <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> to <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> is
unlikely to be needed. Note that conversion from <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a>
to <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> is only possible when all sequences have the
same length, which can be fixed either using alignment utilities or
using the <a class="reference internal" href="data.html#egglib.Container.equalize" title="egglib.Container.equalize"><tt class="xref py py-meth docutils literal"><span class="pre">equalize()</span></tt></a> method.</p>
<p>It is also possible to use <tt class="xref py py-meth docutils literal"><span class="pre">create()</span></tt> to create an instance from
an ad hoc object (provided that iterations returns two- or three-item
tuples:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span> <span class="p">(</span><span class="s">&#39;name1&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name2&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name3&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGTCGCGAGAGCGTTGCCAAGAGTGCGCGTGAT&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name4&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGTCGCGCGCGCGTTGCCAACAGTGCCCGCGAT&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name5&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGTCGCGCGCGCGTTGCCAAGAGTGCCCGCGAT&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name6&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGTCGCGCGCGCGATGCCAAGAGTGCCCGCGAT&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
</pre></div>
</div>
<p>Remember that sequences must have the same length to be imported as an
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span> <span class="p">(</span><span class="s">&#39;name1&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCG&#39;</span><span class="p">),</span>
<span class="gp">... </span>              <span class="p">(</span><span class="s">&#39;name2&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT&#39;</span><span class="p">)</span> <span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">container</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Container</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;Container created, length:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">container</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;Align created, length:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre>Container created, length: 2
Traceback (most recent call last):
    ... traceback omitted ...
ValueError: Sequence doesn&#39;t match the alignment length
</pre></div>
</div>
</div>
<div class="section" id="exports-sequence-data">
<h3>Exports sequence data<a class="headerlink" href="#exports-sequence-data" title="Permalink to this headline">¶</a></h3>
<p>Default format for sequence data is also Pearson&#8217;s FASTA format. FASTA
strings are automatically generated using the <tt class="docutils literal"><span class="pre">print</span></tt> statement (the
<tt class="docutils literal"><span class="pre">align</span></tt> object used here is the same as the last but one example):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">align</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;name1
GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT
&gt;name2
GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT
&gt;name3
GAGCGTGTCGCGAGAGCGTTGCCAAGAGTGCGCGTGAT
&gt;name4
GAGCGTGTCGCGCGCGCGTTGCCAACAGTGCCCGCGAT
&gt;name5
GAGCGTGTCGCGCGCGCGTTGCCAAGAGTGCCCGCGAT
&gt;name6
GAGCGTGTCGCGCGCGCGATGCCAAGAGTGCCCGCGAT
</pre></div>
</div>
<p>This doesn&#8217;t allow to export group labels. To do so, use the
<tt class="xref py py-meth docutils literal"><span class="pre">str()</span></tt> method, which take a <tt class="docutils literal"><span class="pre">exportGroupLabels</span></tt> option (and is
available on both <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">string</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">exportGroupLabels</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">string</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;name1@0
GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT
&gt;name2@0
GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT
&gt;name3@0
GAGCGTGTCGCGAGAGCGTTGCCAAGAGTGCGCGTGAT
&gt;name4@1
GAGCGTGTCGCGCGCGCGTTGCCAACAGTGCCCGCGAT
&gt;name5@1
GAGCGTGTCGCGCGCGCGTTGCCAAGAGTGCCCGCGAT
&gt;name6@1
GAGCGTGTCGCGCGCGCGATGCCAAGAGTGCCCGCGAT
</pre></div>
</div>
<p>To export data directly to a file, use the <tt class="xref py py-meth docutils literal"><span class="pre">write()</span></tt> method (also
available on both <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>).
This method calls underlying C++ code and, on large datasets, is far
more efficient than generating the full string and placing it to a file.
The below example generates a file containing the string of the last
example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;data3.fas&quot;</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>In particular in the case of <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instances, output
in alternative formats can be performed using <a class="reference internal" href="data.html#egglib.Align.nexus" title="egglib.Align.nexus"><tt class="xref py py-meth docutils literal"><span class="pre">nexus()</span></tt></a>,
<a class="reference internal" href="data.html#egglib.Align.phylip" title="egglib.Align.phylip"><tt class="xref py py-meth docutils literal"><span class="pre">phylip()</span></tt></a> and <a class="reference internal" href="data.html#egglib.Align.phyml" title="egglib.Align.phyml"><tt class="xref py py-meth docutils literal"><span class="pre">phyml()</span></tt></a> methods.
Conversely, a few alternative alignment formats are available for input
in the <a class="reference internal" href="tools.html#module-egglib.tools" title="egglib.tools"><tt class="xref py py-mod docutils literal"><span class="pre">tools</span></tt></a> module (see the section &#8220;Specific data
formats&#8221;).</p>
</div>
</div>
<div class="section" id="manipulation-of-sequence-sets-and-alignments">
<h2>Manipulation of sequence sets and alignments<a class="headerlink" href="#manipulation-of-sequence-sets-and-alignments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="iteration">
<h3>Iteration<a class="headerlink" href="#iteration" title="Permalink to this headline">¶</a></h3>
<p>Iteration is frequently useful, in particular for sequence data. When
you iterate over a <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> or a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>
instance, each iteration step yields a <tt class="xref py py-class docutils literal"><span class="pre">SequenceItem</span></tt>
object. You should never need to instanciate a <tt class="xref py py-class docutils literal"><span class="pre">SequenceItem</span></tt>
yourself, and using it should be as simple as accessing it three
attributes <tt class="xref py py-attr docutils literal"><span class="pre">name</span></tt>, <tt class="xref py py-attr docutils literal"><span class="pre">sequence</span></tt> and <tt class="xref py py-attr docutils literal"><span class="pre">group</span></tt>. This class
is designed to allow access/modify the instance without necessarily
copy the sequence data. In the current implementation, the sequence data
will still be extracted if <tt class="xref py py-attr docutils literal"><span class="pre">sequence</span></tt> is
accessed, however. Note that <tt class="xref py py-class docutils literal"><span class="pre">SequenceItem</span></tt> instances
keep a record of their parent <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> or
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance and that, preferably, they should not be
stored outside the loop.</p>
<p>Assume you have a FASTA file <tt class="docutils literal"><span class="pre">data4.fas</span></tt> with an alternative way of
coding population membership (underscore instead of <tt class="docutils literal"><span class="pre">&#64;</span></tt>):</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;name1_1
GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT
&gt;name2_1
GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT
&gt;name3_1
GAGCGTGTCGCGAGAGCGTTGCCAAGAGTGCGCGTGAT
&gt;name4_2
GAGCGTGTCGCGCGCGCGTTGCCAACAGTGCCCGCGAT
&gt;name5_2
GAGCGTGTCGCGCGCGCGTTGCCAAGAGTGCCCGCGAT
&gt;name6_2
GAGCGTGTCGCGCGCGCGATGCCAAGAGTGCCCGCGAT
</pre></div>
</div>
<p>Here is how to import these labels and modify the underlying instance:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="p">(</span><span class="s">&quot;data4.fas&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">align</span><span class="p">:</span>
<span class="gp">... </span>   <span class="n">root</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;_&#39;</span><span class="p">)</span>
<span class="gp">... </span>   <span class="n">label</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
<span class="gp">... </span>   <span class="n">i</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">root</span>
<span class="gp">... </span>   <span class="n">i</span><span class="o">.</span><span class="n">group</span> <span class="o">=</span> <span class="n">label</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;data5.fas&quot;</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>This generates the following file:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;name1@1
GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT
&gt;name2@1
GAGCGTGCCGCGAGAGCGTTGCCAAGATTGCGCGTGAT
&gt;name3@1
GAGCGTGTCGCGAGAGCGTTGCCAAGAGTGCGCGTGAT
&gt;name4@2
GAGCGTGTCGCGCGCGCGTTGCCAACAGTGCCCGCGAT
&gt;name5@2
GAGCGTGTCGCGCGCGCGTTGCCAAGAGTGCCCGCGAT
&gt;name6@2
</pre></div>
</div>
<p>The syntax <tt class="docutils literal"><span class="pre">for</span> <span class="pre">name,</span> <span class="pre">sequence,</span> <span class="pre">group</span> <span class="pre">in</span> <span class="pre">align</span></tt> is supported but has
two flaws: 1) it will cause a performance overhead in case sequences are
long and/or numerous and that they don&#8217;t need to be accessed, and 2)
this doesn&#8217;t allow to modify the underlying instances, since <tt class="docutils literal"><span class="pre">name</span></tt>,
<tt class="docutils literal"><span class="pre">sequence</span></tt> and <tt class="docutils literal"><span class="pre">group</span></tt> variables are strings and integer, and
therefore not modifiable.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When setting the <tt class="xref py py-attr docutils literal"><span class="pre">sequence</span></tt>
attribute of a <tt class="xref py py-class docutils literal"><span class="pre">SequenceItem</span></tt>, the new sequence
must match the length of the alignment.</p>
</div>
</div>
<div class="section" id="sequence-data-edition">
<h3>Sequence data edition<a class="headerlink" href="#sequence-data-edition" title="Permalink to this headline">¶</a></h3>
<p>It is possible to access (read or modify) full data at a given position
using subscript indexing (<tt class="docutils literal"><span class="pre">[]</span></tt>). The operator takes or returns a
three-item tuple:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;first name&#39;</span><span class="p">,</span> <span class="s">&#39;AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&#39;</span><span class="p">,</span> <span class="mi">8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;first name&#39;</span><span class="p">,</span> <span class="s">&#39;TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">&#39;name1&#39;</span><span class="p">,</span> <span class="s">&#39;GAGCGTGCCGCGAGAGCGTTGCCAAGAGTGCCCGTGAT&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="p">(</span><span class="s">&#39;first name&#39;</span><span class="p">,</span> <span class="s">&#39;AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&#39;</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="p">(</span><span class="s">&#39;first name&#39;</span><span class="p">,</span> <span class="s">&#39;TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that, if the group label is omitted, the value of <tt class="docutils literal"><span class="pre">0</span></tt> is assumed
and replaces the previous value.</p>
<p>It is possible to use negative indices, as in <tt class="docutils literal"><span class="pre">align[-1]</span></tt>. The <tt class="docutils literal"><span class="pre">del</span></tt>
statement is allowed for removing a full entry (name, sequence and group)
as in <tt class="docutils literal"><span class="pre">del</span> <span class="pre">align[2]</span></tt>. To remove a sequence based on its name, use
<tt class="xref py py-meth docutils literal"><span class="pre">remove()</span></tt>.</p>
<p>It is never allowed to access indices equal to or larger than the
current length of the sequence set or alignment, including to add a
new sequence (see below). The methods <a class="reference internal" href="data.html#egglib.Align.set" title="egglib.Align.set"><tt class="xref py py-meth docutils literal"><span class="pre">set()</span></tt></a> and
<a class="reference internal" href="data.html#egglib.Align.get" title="egglib.Align.get"><tt class="xref py py-meth docutils literal"><span class="pre">get()</span></tt></a> allow
to access a single character at a specific position. Note also that
the overloaded methods <tt class="xref py py-meth docutils literal"><span class="pre">name()</span></tt>,
<tt class="xref py py-meth docutils literal"><span class="pre">sequence()</span></tt> and <tt class="xref py py-meth docutils literal"><span class="pre">group()</span></tt>
allow to access/modify only one of these attribute directly at a given
position. <tt class="xref py py-meth docutils literal"><span class="pre">sequenceByName()</span></tt> allows to get a sequence
string using the name and not the index.</p>
<p>It is possible to add a stretch a sequence to a sequence at a given
position, on <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> only, using the method
<tt class="xref py py-meth docutils literal"><span class="pre">appendSequence()</span></tt>.</p>
<p>The method <tt class="xref py py-meth docutils literal"><span class="pre">append()</span></tt> allows to add a new sequence at
the end of the sequence set or alignment, and the method
<tt class="xref py py-meth docutils literal"><span class="pre">addSequences()</span></tt> allows to do it repeatively.</p>
<p>On a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance, the methods
<a class="reference internal" href="data.html#egglib.Align.column" title="egglib.Align.column"><tt class="xref py py-meth docutils literal"><span class="pre">column()</span></tt></a> and <tt class="xref py py-meth docutils literal"><span class="pre">removePosition()</span></tt>
allow to (respectively) access and remove a full column.</p>
<p>Finally, the code below allow to extract the first three sequences of
an alignment, and only the first 10 and the last 10 positions:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">part</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">part</span> <span class="o">=</span> <span class="n">part</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span><span class="o">+</span><span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">part</span>
</pre></div>
</div>
<p>results in:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;name1
GAGCGTGCCGTGCCCGTGA
&gt;name2
GAGCGTGCCGTGCGCGTGA
&gt;name3
GAGCGTGTCGTGCGCGTGA
</pre></div>
</div>
</div>
</div>
<div class="section" id="manipulation-of-other-data-types-trees-annotated-sequences-ssr">
<h2>Manipulation of other data types (trees, annotated sequences, SSR)<a class="headerlink" href="#manipulation-of-other-data-types-trees-annotated-sequences-ssr" title="Permalink to this headline">¶</a></h2>
<p>Phylogenetic trees are implemented as the class <a class="reference internal" href="data.html#egglib.Tree" title="egglib.Tree"><tt class="xref py py-class docutils literal"><span class="pre">Tree</span></tt></a>
and GenBank flatfile data as the class <tt class="xref py py-class docutils literal"><span class="pre">GenBank</span></tt>. Both
can also be instanciated from files and strings. Similarly to sequence
sets and alignments, iteration over instances of these two classes
yield ad hoc class instances allowing to manage, respectively, a node of
the tree and a feature. Available methods on trees allow to edit the
them (such as rooting) and explore them (such as identifying monophyletic
groups), while methods on GenBank flatfile data allow to access and
edit annotation information and extract data (such as coding sequence
strings). The <tt class="xref py py-class docutils literal"><span class="pre">SSR</span></tt> class is designed to import, export
and compute statistics on microsatellite data.</p>
</div>
<div class="section" id="analysis-of-polymorphism">
<h2>Analysis of polymorphism<a class="headerlink" href="#analysis-of-polymorphism" title="Permalink to this headline">¶</a></h2>
<p>Function to compute polymorphism are available as methods of the
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> and <tt class="xref py py-class docutils literal"><span class="pre">SSR</span></tt> classes:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="xref py py-meth docutils literal"><span class="pre">polymorphism()</span></tt>: compute standard statistics,
neutrality indices, and differentiation statistics (if applicable).</li>
<li><tt class="xref py py-meth docutils literal"><span class="pre">polymorphismBPP()</span></tt>: compute statistics
available through the Bio++ library (including statistics based on
nonsynonymous/synonymous variation, if applicable). Bio++ must
have been available when installing EggLib to allow using this
function.</li>
<li><tt class="xref py py-meth docutils literal"><span class="pre">stats()</span></tt>: compute microsatellite statistics.</li>
<li><tt class="xref py py-meth docutils literal"><span class="pre">Fstats()</span></tt>: compute F-statistics (based on Weir
and Cockerham formulas) for microsatellite.</li>
</ul>
<p>Note that the first three methods return dictionaries and that, in
the case of the first two methods, the content of these dictionaries
will depend on both the value of options (some statistics can be
skipped) and on the content of the instance (differentiation
statistics are not computed if only one population is present).
Please refer to the documentation of these methods for more details.</p>
</div></blockquote>
</div>
<div class="section" id="coalescent-simulations">
<h2>Coalescent simulations<a class="headerlink" href="#coalescent-simulations" title="Permalink to this headline">¶</a></h2>
<p>The structure of the <a class="reference internal" href="simul.html#module-egglib.simul" title="egglib.simul"><tt class="xref py py-mod docutils literal"><span class="pre">simul</span></tt></a> module (see its description)
is designed as a compromise between flexibility (make available all
features of the underlying C++ library), usability (not being too clumsy
with a large number of options in a single function) and simplicity
(reduce the number of required steps). The below example shows how to
perform a given number of simulations of a moderately complex model
(two populations that of which one is the result of a founding event
from the first one, and a standard model of mutation with an infinite
number of sites). Note that many more elaborate scenarios and mutation
models are available.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># sets parameters</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample1</span> <span class="o">=</span> <span class="mi">40</span>   <span class="c"># number of sample in pop 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample2</span> <span class="o">=</span> <span class="mi">40</span>   <span class="c"># number of sample in pop 2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">date1</span> <span class="o">=</span> <span class="mf">0.20</span>   <span class="c"># end of bottleneck</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">date2</span> <span class="o">=</span> <span class="mf">0.25</span>   <span class="c"># split date/start of bottleneck</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">strength</span> <span class="o">=</span> <span class="mf">0.1</span> <span class="c"># strength of the bottleneck</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">migr</span> <span class="o">=</span> <span class="mf">0.0</span>     <span class="c"># migration rate (default is 0.1)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">theta</span> <span class="o">=</span> <span class="mf">2.</span>     <span class="c"># theta for gene</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nrepets</span> <span class="o">=</span> <span class="mi">1000</span> <span class="c"># number of simulations</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># sets the two parameter classes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paramSet</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">simul</span><span class="o">.</span><span class="n">CoalesceParamSet</span><span class="p">([</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">],</span> <span class="n">M</span><span class="o">=</span><span class="n">migr</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paramSet</span><span class="o">.</span><span class="n">changeSinglePopulationSize</span><span class="p">(</span><span class="n">date1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">strength</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paramSet</span><span class="o">.</span><span class="n">populationFusion</span><span class="p">(</span><span class="n">date2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mutator</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">simul</span><span class="o">.</span><span class="n">CoalesceFiniteAlleleMutator</span><span class="p">(</span><span class="n">theta</span><span class="o">=</span><span class="mf">2.</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># performs simulations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligns</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">simul</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">paramSet</span><span class="p">,</span> <span class="n">mutator</span><span class="p">,</span> <span class="n">nrepets</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># computes statistics</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pol_data</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">egglib</span><span class="o">.</span><span class="n">Align</span><span class="o">.</span><span class="n">polymorphism</span><span class="p">,</span> <span class="n">aligns</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">D</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="s">&#39;D&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pol_data</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Fst</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="s">&#39;Fst&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pol_data</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;average D:&#39;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">D</span><span class="p">)</span><span class="o">/</span><span class="n">nrepets</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="s">&#39;average Fst:&#39;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">Fst</span><span class="p">)</span><span class="o">/</span><span class="n">nrepets</span>
</pre></div>
</div>
<p>Note that is is not necessary to formally end the bottleneck by
<tt class="docutils literal"><span class="pre">paramSet.changeSinglePopulationSize(date0,</span> <span class="pre">1,</span> <span class="pre">1.)</span></tt> since this
population is removed at the same date.</p>
<p>Here&#8217;s the result of a particular run:</p>
<div class="highlight-python"><div class="highlight"><pre>average D: -0.557003767734
average Fst: 0.373726230187
</pre></div>
</div>
</div>
<div class="section" id="interactive-usage-and-approximate-bayesian-computation">
<h2>Interactive usage and Approximate Bayesian Computation<a class="headerlink" href="#interactive-usage-and-approximate-bayesian-computation" title="Permalink to this headline">¶</a></h2>
<p>In this section we describe the command line tools available through
EggLib. A number of short programs have been written under the form
of subclasses of <a class="reference internal" href="utils.html#egglib.utils.BaseCommand" title="egglib.utils.BaseCommand"><tt class="xref py py-class docutils literal"><span class="pre">BaseCommand</span></tt></a> in the module
<a class="reference internal" href="utils.html#module-egglib.utils" title="egglib.utils"><tt class="xref py py-mod docutils literal"><span class="pre">utils</span></tt></a>. The <tt class="docutils literal"><span class="pre">egglib</span></tt> script (which is installed at the
same tyme as the EggLib Python module), lets you run automatically these
programs. In this section we present some generalities on the general
usage of the <tt class="docutils literal"><span class="pre">egglib</span></tt> script, and then provide an example of usage
with a tutorial for ABC analyses using such commands. Note that all
commands presented below are system commands (entered through a command
terminal) and not Python code.</p>
<div class="section" id="the-egglib-script">
<h3>The egglib script<a class="headerlink" href="#the-egglib-script" title="Permalink to this headline">¶</a></h3>
<p>The general usage of the <tt class="docutils literal"><span class="pre">egglib</span></tt> script is:</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib &lt;command&gt; &lt;options&gt;
</pre></div>
</div>
<p>where <em>command</em> is the name of any of the available programs. An
alternative syntax is available as <tt class="docutils literal"><span class="pre">python</span> <span class="pre">-m</span> <span class="pre">egglib.utils</span> <span class="pre">&lt;command&gt;</span> <span class="pre">&lt;options&gt;</span></tt></p>
<p>By typing</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib
</pre></div>
</div>
<p>only, a list of all commands will be displayed. By typing the command
name only (without options), you will be provided with the manual page
of this command. For example in the case of one of the simplest commands,
<em>infos</em>:</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib infos
</pre></div>
</div>
<p>Concerning <em>options</em>, they are variables among commands (the command&#8217;s
manual page always provide the list and documentation). It is important
to note that, even if the syntax is standardized over different commands,
they really are different programs with different syntaxes (see however
the <em>quiet</em> and <em>debug</em> boolean flags below. Options can be of two types:</p>
<blockquote>
<div><ul>
<li><p class="first"><strong>Keyword arguments</strong>. They are keyword-identified values, entered in
the syntax <tt class="docutils literal"><span class="pre">keyword=value</span></tt> where <em>keyword</em> is one of the
available options and <em>value</em> the value you wish to enter. They
are usually some restrictions on the format of the value and some
options are flexible. In some cases (especially for values that
contain characters that can be mistaken with command interpreter
modifiers such a spaces and semicolons) it can be necessary to
wrap the option in quotes. Some keyword arguments have a default
value, and some have not. The latter must be entered to run the
command. The <em>input</em> option is found in most commands, and
frequently expects a file name as in <tt class="docutils literal"><span class="pre">input=file</span></tt> (allowing
sometimes to enter multiple file names as in
<tt class="docutils literal"><span class="pre">input=file1,file2,file3</span></tt>), but in a few cases it expects the
name of a directory containing files to process.</p>
</li>
<li><p class="first"><strong>Boolean flags</strong>. They are entered as a single token and they
activate a specified feature. They are never required and are
always off by default (but note that some flags have a negative
effect, such a <tt class="docutils literal"><span class="pre">no_check</span></tt> in command <tt class="docutils literal"><span class="pre">cprimers</span></tt>). Two flags
are available automatically for all commands:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">quiet</span></tt>: If this flag is activated, the command will run
without console output. Note that the flag has not effect on
commands that don&#8217;t generate console output anyway, as well as
commands whose only task is to generate console output (this
is the case for the <tt class="docutils literal"><span class="pre">infos</span></tt> command, for example).</li>
<li><tt class="docutils literal"><span class="pre">debug</span></tt>: By default, all errors are intercepted by the
<em>egglib</em> script and summarized in a one-line error statement.
The flag <tt class="docutils literal"><span class="pre">debug</span></tt> allows to output the full error traceback
(this is useful for debugging and reporting errors).</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
<p>Options (keyword arguments as well as boolean flags) can be entered in
any order, but cannot be repeated. It is always required to enter
at least one keyword argument (sometimes, several must be entered).</p>
<p>Here is the end of the example with the <tt class="docutils literal"><span class="pre">infos</span></tt> command. Here is how
to get information on the <tt class="docutils literal"><span class="pre">data4.fas</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib infos input=data4.fas
data4.fas
 ... 6 sequences
 ... alignment
 ... length: 38
</pre></div>
</div>
<p>And how to get information on several files (<tt class="docutils literal"><span class="pre">data2.txt</span></tt> contains
names but no sequences):</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib infos input=data1.fas,data2.fas,data3.fas,data4.fas
data1.fas
 ... 10 sequences
 ... alignment
 ... length: 160
data2.fas
 ... 10 sequences
 ... alignment
 ... length: 0
data3.fas
 ... 6 sequences
 ... alignment
 ... length: 38
data4.fas
 ... 6 sequences
 ... alignment
 ... length: 38
</pre></div>
</div>
<p>The default error statement (the file <tt class="docutils literal"><span class="pre">data6.fas</span></tt> doesn&#8217;t exist):</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib infos input=data6.fas
An error occurred: [ValueError] cannot open data6.fas
</pre></div>
</div>
<p>And how to get more information:</p>
<div class="highlight-python"><div class="highlight"><pre>$ egglib infos input=data6.fas debug
An error occurred: [ValueError] cannot open data6.fas
Traceback (most recent call last):
  File &quot;/usr/bin/egglib&quot;, line 67, in &lt;module&gt;
    obj.run(*args, **kwargs)
  File &quot;/usr/lib64/python2.7/site-packages/egglib/utils.py&quot;, line 209, in run
    self._run(*fargs, **kwargs)
  File &quot;/usr/lib64/python2.7/site-packages/egglib/utils.py&quot;, line 3386, in _run
    container = data.Container(fname)
  File &quot;/usr/lib64/python2.7/site-packages/egglib/data.py&quot;, line 767, in __init__
    if not os.path.isfile(fname): raise ValueError, &#39;cannot open %s&#39; %fname
ValueError: cannot open data6.fas
</pre></div>
</div>
<p>(Of course in that case the source of the error can be identified with
the message alone. In that case, the exception has been thrown by the
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> constructor when the file was not found.)</p>
</div>
<div class="section" id="abc-sampling">
<h3>ABC sampling<a class="headerlink" href="#abc-sampling" title="Permalink to this headline">¶</a></h3>
<p>In this example we will consider a simulated dataset of 10 independent
alignments of 1000 nucleotides, simulated under a model containing
2 isolated populations of equal size and connected with a bidirectional
migration rate (4Nm) of 0.5, with a mutation rate (4Nu) of 3.5 for gene
(that is 0.0035 per site). 20 sequences were sampled in each population.
We will see if it is possible to reject the standard neutral model (SNM)
in favour of the (real) island model with, in that case, two demes (IM)
and if it is possible to correctly estimate the parameters.</p>
<p>We will first sample random parameters in uniform priors under these two
alternative models (SNM and IM) and perform one simulation for each
sample (one simulation contains the 10 loci). The command to do so is:</p>
<div class="highlight-python"><div class="highlight"><pre>egglib abc_sample dir=fas params=SNM.txt data=SNM.out model=SNM prior=&quot;%U(0;0.01)&quot; stats=SFS:4
</pre></div>
</div>
<p>The following arguments were entered:</p>
<blockquote>
<div><ul class="simple">
<li><em>dir</em>: The name of a directory containing fasta files to analyze.</li>
<li><em>params</em>: The name of the run parameter output file.</li>
<li><em>data</em>: The name of data output file.</li>
<li><em>model</em>: The label of the demographic model, here the Standard
Neutral Model.</li>
<li><em>prior</em>: The definition of the prior. Here, the prior is defined
as a string (starting with a <tt class="docutils literal"><span class="pre">%</span></tt> character to indicate that the
argument is not a file name), and specifies a uniform distribution
bound between 0 and 0.01 (the SNM model has only one parameter).</li>
<li><em>stats</em>: The set of summary statistics, here the Site Frequency
Spectrum. This particular set of summary statistics expects an
argument (the number of classes in the spectrum), which is passed
after the <tt class="docutils literal"><span class="pre">:</span></tt> character.</li>
</ul>
</div></blockquote>
<p>The following parameters were left to the default values:</p>
<blockquote>
<div><ul class="simple">
<li><em>ext</em> (<tt class="docutils literal"><span class="pre">fas</span></tt>): This argument allows to set the extension of
fasta files to process.</li>
<li><em>post</em> (10000): The number of point to sample for posterior
estimation.</li>
<li><em>step</em> (100): The rate of refreshing of the progress information.</li>
<li><em>seeds</em> (automatic): Random number generator seeds.</li>
<li><em>restart</em>: This option should be used to resume an interrupted. It
takes a run parameter file (the file named by the <em>params</em> option)
and will attempt to finish the run. No other options should be
entered along with <em>restart</em>.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="abc-rejection-regression-procedure">
<h3>ABC rejection-regression procedure<a class="headerlink" href="#abc-rejection-regression-procedure" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">abc_fit</span></tt> command performs the steps of the standard
rejection-regression procedure described in Beaumont et al. 2002 Genetics
using the underlying C++ class <a class="reference internal" href="binding.html#egglib.egglib_binding.ABC" title="egglib.egglib_binding.ABC"><tt class="xref py py-class docutils literal"><span class="pre">ABC</span></tt></a>:</p>
<div class="highlight-python"><div class="highlight"><pre>egglib abc_fit input=SNM.txt tolerance=0.1 transform=tan output=SNM.fit
</pre></div>
</div>
<p>Here we filled all arguments. The input file is <tt class="docutils literal"><span class="pre">SNM.txt</span></tt> which
actually contains only the values of the run parameters. The actual data
file is given as one of the parameters and must be found in the same
directory. The tolerance gives the proportion of points to select in
the region closest to the observed statistics. This defines the local
region and, in practice, should be more stringent than in this example.</p>
<blockquote>
<div>The fitting procedure can be applied to data from a unique model only.</div></blockquote>
<p>The transformation of parameter values (here, tangential transformation)
is useful in particular to prevent parameters from exceeding prior bounds.
The output file will contain only parameter values, selected and corrected,
and will be considered as a posterior distribution.</p>
</div>
<div class="section" id="comparing-abc-models">
<h3>Comparing ABC models<a class="headerlink" href="#comparing-abc-models" title="Permalink to this headline">¶</a></h3>
<p>The rejection-based method of Fagundes et al. 2007 PNAS allows to
compare alternative model without relying on Bayes Factors. It is based
on the rejection step of the rejection-regression procedure, which
doesn&#8217;t use the parameter values and therefore allows to merge several
models. The model probabilities are estimated as the proportion of
points originating from each model in a pre-defined number of accepted
points:</p>
<div class="highlight-python"><div class="highlight"><pre>egglib abc_compare input=SNM.txt,PEM.txt tolerance=0.1 output=compare.txt
</pre></div>
</div>
<p>The different sample files are passed in the <em>input</em> argument. The
<em>tolerance</em> argument is the same as for the fitting procedure, but note
that here it applies to the sum of points over all the passed models.
The output file will contain the accepted points, along with their
weigths. The model probabilities will be displayed in the standard
output (the terminal).</p>
</div>
<div class="section" id="analyzing-posterior-distribution">
<h3>Analyzing posterior distribution<a class="headerlink" href="#analyzing-posterior-distribution" title="Permalink to this headline">¶</a></h3>
<p>A set of commands are designed to help analyzing posteriors, after
they have been fitted.</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">abc_bin</span></tt> discretizes posteriors into a regular grid over all
parameter dimensions.</li>
<li><tt class="docutils literal"><span class="pre">abc_statsmarg</span></tt> provide statistics for individual parameters
independently, from a fitted posterior.</li>
<li><tt class="docutils literal"><span class="pre">abc_statsdisc</span></tt> (for models with more than one parameter),
analyzes a discretized posterior (and, in particular, identifies
the maximum density point of the distribution).</li>
<li><tt class="docutils literal"><span class="pre">abc_plot1D</span></tt> and <tt class="docutils literal"><span class="pre">abc_plot2D</span></tt> display graphical representations
parameters (respectively, a single parameter and the covariation
of two parameters). They require that <tt class="xref py py-mod docutils literal"><span class="pre">matplotlib</span></tt> is
available and they use discretized posteriors.</li>
<li><tt class="docutils literal"><span class="pre">abc_psimuls</span></tt>, finally, allows to use posteriors to generate
new simulations for testing neutrality of loci or assess
goodness-of-fit of models.</li>
</ul>
</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="eggcoal.html" title="eggcoal"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Download and install"
             >previous</a> |</li>
        <li><a href="index.html">EggLib</a> &raquo;</li> 
      </ul>
    </div>
        <address style="text-align: right;"><small>Hosted by&nbsp;
        <a href="http://sourceforge.net/projects/seqlib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get seqlib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
        </address>

  </body>
</html>