<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>tools &mdash; EggLib</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.1.11',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/icon.ico"/>
    <link rel="top" title="EggLib" href="index.html" />
    <link rel="up" title="Python module" href="egglib-py.html" />
    <link rel="next" title="wrappers" href="wrappers.html" />
    <link rel="prev" title="Top-level Egglib components" href="data.html" /> 
  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px 0px 0px 0px">
<a href="index.html"><img src="_static/egglib_banniere5.png" border="0" alt="EggLib"/></a>
</div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="wrappers.html" title="wrappers"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="data.html" title="Top-level Egglib components"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">EggLib</a> &raquo;</li>
          <li><a href="egglib-py.html" accesskey="U">Python module</a> &raquo;</li> 
      </ul>
    </div>
<div style="background-color: grey; text-align: left; padding: 0px 0px 0px 10px; border-bottom: 1px solid #2d2e91">
<font color="white" ><b>EggLib version 2.1 is archived.</b></font></div>


      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">tools</a><ul>
<li><a class="reference internal" href="#specific-data-formats">Specific data formats</a></li>
<li><a class="reference internal" href="#data-analysis">Data analysis</a></li>
<li><a class="reference internal" href="#sequence-manipulation-tools">Sequence manipulation tools</a></li>
<li><a class="reference internal" href="#sequence-comparison">Sequence comparison</a></li>
<li><a class="reference internal" href="#tools-not-directly-related-to-sequence-data">Tools not (directly) related to sequence data</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="data.html"
                        title="previous chapter">Top-level Egglib components</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="wrappers.html"
                        title="next chapter">wrappers</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/tools.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-egglib.tools">
<span id="tools"></span><h1>tools<a class="headerlink" href="#module-egglib.tools" title="Permalink to this headline">¶</a></h1>
<p>This module contains tools used in some other parts of EggLib but that
might be of use for the package&#8217;s users.</p>
<div class="section" id="specific-data-formats">
<h2>Specific data formats<a class="headerlink" href="#specific-data-formats" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="egglib.tools.aln2fas">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">aln2fas</tt><big>(</big><em>fname</em><big>)</big><a class="headerlink" href="#egglib.tools.aln2fas" title="Permalink to this definition">¶</a></dt>
<dd><p>Imports a clustal-formatted alignment from the file name <em>fname</em> and
returns a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.staden">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">staden</tt><big>(</big><em>fname=None</em>, <em>string=None</em>, <em>delete_consensus=True</em><big>)</big><a class="headerlink" href="#egglib.tools.staden" title="Permalink to this definition">¶</a></dt>
<dd><p>Imports a Staden output file as an <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance. The
file should have been  generated from a contig alignment by the GAP4
contig editor, using the command &#8220;dump contig. to file&#8221;. The
sequence named <tt class="docutils literal"><span class="pre">CONSENSUS</span></tt>, if present, is automatically removed
unless the option <em>delete_consensus</em> is <tt class="docutils literal"><span class="pre">False</span></tt>.</p>
<p>The Staden outfile file can be read from a file (using the argument
<em>fname</em> or directly from a string (using <em>string</em>). It is required
to pass either a file name as <em>fname</em> or a Staden string as <em>string</em>
and it is not allowed to passb both.</p>
<p>Staden&#8217;s default convention is followed:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">-</span></tt> codes for an unknown base and is replaced by <tt class="docutils literal"><span class="pre">N</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">*</span></tt> codes for an alignment gap and is replaced by <tt class="docutils literal"><span class="pre">-</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">.</span></tt> represents the same sequence than the consensus at that
position.</li>
<li>White space represents missing data and is replaced by <tt class="docutils literal"><span class="pre">?</span></tt>.</li>
</ul>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1: </span>Add argument <em>delete_consensus</em>.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.1.0: </span>Read from string or fname.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.get_fgenesh">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">get_fgenesh</tt><big>(</big><em>fname</em><big>)</big><a class="headerlink" href="#egglib.tools.get_fgenesh" title="Permalink to this definition">¶</a></dt>
<dd><p>Imports fgenesh output. <em>fname</em> must be the name of a file
containing fgenesh ouput. The feature definition are parsed an
returned as a list of <tt class="docutils literal"><span class="pre">gene</span></tt> and <tt class="docutils literal"><span class="pre">CDS</span></tt> features represented by
dictionaries. Note that 5&#8217; partial features might not be in the
appropriate frame and that it can be necessary to add a <tt class="docutils literal"><span class="pre">codon_start</span></tt>
qualifier.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.genalys2fasta">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">genalys2fasta</tt><big>(</big><em>iname</em><big>)</big><a class="headerlink" href="#egglib.tools.genalys2fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts Genalys-formatted sequence alignment files to fasta. The
function imports files generated through the option <em>Save SNPs</em> of
Genalys 2.8. <em>iname</em> if the name of the Genalys output file. Returns
an <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance.</p>
</dd></dl>

<dl class="class">
<dt id="egglib.tools.Mase">
<em class="property">class </em><tt class="descclassname">egglib.tools.</tt><tt class="descname">Mase</tt><big>(</big><em>input=None</em><big>)</big><a class="headerlink" href="#egglib.tools.Mase" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">list</span></tt></p>
<p>Minimal implementation of the mase format (allowing input/output
operations). This class emulates a list of dictionaries, each
dictionary representing a sequence and describing the keys <em>header</em>,
<em>name</em> and <em>sequence</em>. However, the string formatter (<tt class="docutils literal"><span class="pre">str(mase)</span></tt>
or <tt class="docutils literal"><span class="pre">print</span> <span class="pre">mase</span></tt>, where <em>mase</em> is a <a class="reference internal" href="#egglib.tools.Mase" title="egglib.tools.Mase"><tt class="xref py py-class docutils literal"><span class="pre">Mase</span></tt></a> instance)
generates a mase-formatted string. Object attributes are <em>header</em>
(a string with file-level information), <em>species</em> (the species of
the ingroup), <em>align</em> (an <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance corresponding
to the data contained in the instance, and created upon construction).
Modifying this instance has no effect.</p>
<p>The constructor takes an optional argument that can be a string
giving the path to a mase-formatted file, or a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a>
instance. The constructor is currently unable to import
population labels, and only sequences marked as <em>ingroup</em> are
imported.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.0.1: </span>An <tt class="xref py py-class docutils literal"><span class="pre">IOError</span></tt> is raised upon file formatting error.</p>
</div>
</dd></dl>

</div>
<div class="section" id="data-analysis">
<h2>Data analysis<a class="headerlink" href="#data-analysis" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="egglib.tools.LD">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">LD</tt><big>(</big><em>align1</em>, <em>align2</em>, <em>shuffle</em><big>)</big><a class="headerlink" href="#egglib.tools.LD" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes linkage disequilibrium statistics between two <tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt>
instances <em>align1</em> and <em>align2</em>. If <em>shuffle</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, randomly
shuffles sequences of <em>align2</em> (without altering the original
instance), emulating the hypothesis of linkage equilibrium. Returns
a <tt class="docutils literal"><span class="pre">(n1,n2,</span> <span class="pre">S1,S2,K1,K2,D,Dp)</span></tt> tuple, where <em>n1</em> is the number of
used sequences of the first alignment, <em>S1</em> is the number of
polymorphic sites of the first alignment, <em>K1</em> is the number of
unique haplotypes of the first alignment, <em>D</em> is the standard
estimator of linkage disequilibrium and <em>Dp</em> is Lewontin&#8217;s estimator
(bound by 0 and 1).</p>
</dd></dl>

</div>
<div class="section" id="sequence-manipulation-tools">
<h2>Sequence manipulation tools<a class="headerlink" href="#sequence-manipulation-tools" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="egglib.tools.backalign">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">backalign</tt><big>(</big><em>nucseq</em>, <em>protseq</em>, <em>code=1</em>, <em>has_stop=False</em><big>)</big><a class="headerlink" href="#egglib.tools.backalign" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignement of coding sequences based on aligned predicted products.
Conceptual translations of DNA sequences must match exactly passed
protein sequences (except for gaps). Stop codons are not supported.
<em>nucseq</em> is a <a class="reference internal" href="data.html#egglib.Container" title="egglib.Container"><tt class="xref py py-class docutils literal"><span class="pre">Container</span></tt></a> instance containing raw
coding sequence. <em>protseq</em> is a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance
containing aligned amino acid sequences. <em>code</em> specifies the genetic
code; refer to the documentation of <a class="reference internal" href="#egglib.tools.GeneticCodes.translate" title="egglib.tools.GeneticCodes.translate"><tt class="xref py py-meth docutils literal"><span class="pre">translate()</span></tt></a>.
Returns a <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance containing aligned coding
sequences. <em>has_stop</em> nucseq sequences are allowed to contain a
final stop codon.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.concat">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">concat</tt><big>(</big><em>aligns</em>, <em>spacer=0</em>, <em>ch='?'</em>, <em>strict=True</em>, <em>groupCheck=True</em><big>)</big><a class="headerlink" href="#egglib.tools.concat" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates sequence alignments. A unique <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> is
returned. All different sequences from all passed alignments are
represented in the final alignment. Sequences whose name match are
matching are concatenated. In case several sequences have the same
name in a given segment, the first one is considered and others are
discarded. In case a sequence is missing for a particular segment,
a stretch of non-varying characters is inserted to replace the
unknown sequence.</p>
<p><em>aligns</em> must be an iterable containing <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instances.</p>
<p><em>spacer</em> specifies the length  of  unsequenced stretches
(represented by non-varying characters)  between concatenated
alignments. If <em>spacer</em> is a positive integer, the length of all
stretches will be identical. If <em>spacer</em> is an iterable containing
integers, each specifying the interval between two consecutive
alignments (if <em>aligns</em> contains <tt class="docutils literal"><span class="pre">n</span></tt> alignments, <em>spacer</em> must be
of length <tt class="docutils literal"><span class="pre">n-1</span></tt>).</p>
<p><em>ch</em> gives the character to used for conserved stretches and for
missing segments.</p>
<p>If <em>strict</em> is <tt class="docutils literal"><span class="pre">False</span></tt>, the name comparison will not extend
further than the length of the shorter name: for example, names
<tt class="docutils literal"><span class="pre">anaconda</span></tt> and <tt class="docutils literal"><span class="pre">anaco</span></tt> will match,  and the concatenated
sequence will be named <tt class="docutils literal"><span class="pre">anaconda</span></tt> (regardless of which name
appears first in the list of <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instances).</p>
<p>If <em>groupCheck</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, an exception will be raised in case of
a mismatch between group labels of different sequence segments
bearing the same name. Otherwise, the group of the first segment
found will be used as group label of the final sequence.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.0.1: </span>The arguments allowing to customize function&#8217;s behaviour. </p>
</div>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.longest_orf">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">longest_orf</tt><big>(</big><em>sequence</em>, <em>clean=False</em>, <em>full=False</em>, <em>all=False</em>, <em>code=1</em>, <em>mini=1</em><big>)</big><a class="headerlink" href="#egglib.tools.longest_orf" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the longest open reading frame in the sequence. By default,
the longest sequence without stop codon (except for the trailing
stop codon) is returned, therefore the returned ORF doesn&#8217;t
necessarily start by ATG and stops by a stop  codon. If <em>clean</em> is
<tt class="docutils literal"><span class="pre">True</span></tt>, returns the longest sequence encoding a valid protein
sequence (without stop, without missing data, without gap). If
<em>full</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, returns only genuine ORFs (starting by ATG and
ending by a stop codon). If <em>all</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, returns a list of all
ORFs (at least 3 of length), sorted by decreasing length. <em>code</em>
specifies the genetic code; refer to the documentation of
<a class="reference internal" href="#egglib.tools.GeneticCodes.translate" title="egglib.tools.GeneticCodes.translate"><tt class="xref py py-meth docutils literal"><span class="pre">GeneticCodes.translate()</span></tt></a>. <em>mini</em> specifies the minimum number
of codons (or amino acids) of the returned ORF or ORFs (stop codons
are not taken into account).</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.0.1: </span>Added options; return the trailing stop codon when appropriate.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.1.0: </span>Added option <em>mini</em>. The behaviour of previous versions is
reproduced by setting <em>mini</em> to 0.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.rc">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">rc</tt><big>(</big><em>seq</em><big>)</big><a class="headerlink" href="#egglib.tools.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse-complements a DNA sequence. Upper and lower-cases characters
can be passed, the output is always upper-case. IUPAC characters
(<tt class="docutils literal"><span class="pre">ACGTMRWSYKBDHV</span></tt>) are reverted. The characters <tt class="docutils literal"><span class="pre">N-?</span></tt> are
returned as is. Other characters raise a ValueError.</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.0.1: </span>Characters N, - and ? are correctly processed.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.0.2: </span>Reimplemented (will be faster for large sequences).</p>
</div>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.translate">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">translate</tt><big>(</big><em>input</em>, <em>code=1</em>, <em>strip=False</em><big>)</big><a class="headerlink" href="#egglib.tools.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translates all sequences from nucleotide to proteins. Accepts
sequence container instances and the return type matches the passed
type. If <em>strip</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, all stop codon(s) present at the end
of any sequence will be automatically stripped off. Setting this
option to <tt class="docutils literal"><span class="pre">True</span></tt> will raise a <tt class="xref py py-class docutils literal"><span class="pre">ValueError</span></tt> in case a
<a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> is passed and sequences don&#8217;t have all the
same number of trailing stop codons. See the documentation of
<a class="reference internal" href="#egglib.tools.GeneticCodes.translate" title="egglib.tools.GeneticCodes.translate"><tt class="xref py py-meth docutils literal"><span class="pre">GeneticCodes.translate()</span></tt></a> for documentation of the argument
<em>code</em>. Ambiguous codons are translated if the implied possibilities
translate all to the same codon. The IUPAC nomenclature is used.
Note that <tt class="docutils literal"><span class="pre">N</span></tt> means <tt class="docutils literal"><span class="pre">A</span></tt>, <tt class="docutils literal"><span class="pre">C</span></tt>, <tt class="docutils literal"><span class="pre">G</span></tt> or <tt class="docutils literal"><span class="pre">T</span></tt> but that codons
containing <tt class="docutils literal"><span class="pre">?</span></tt> or <tt class="docutils literal"><span class="pre">-</span></tt> will  always be translated as <tt class="docutils literal"><span class="pre">X</span></tt>
(except for <tt class="docutils literal"><span class="pre">---</span></tt> codons that are be translated as <tt class="docutils literal"><span class="pre">-</span></tt>).</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.ungap">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">ungap</tt><big>(</big><em>align</em>, <em>freq</em>, <em>includeOutgroup=True</em><big>)</big><a class="headerlink" href="#egglib.tools.ungap" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a new <a class="reference internal" href="data.html#egglib.Align" title="egglib.Align"><tt class="xref py py-class docutils literal"><span class="pre">Align</span></tt></a> instance containing all
sequences of  <em>align</em> and only the columns for which the frequency
of gaps (<tt class="docutils literal"><span class="pre">-</span></tt> symbols) is less than the value given by <em>freq</em>.</p>
<p>If <em>includeOutgroup</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, the sequences with group label 999
(if any) are considered for computing the frequency of gaps. These
sequences are however always exported to the returned alignment).</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.1.0: </span>Added option <em>includeOutgroup</em>.</p>
</div>
</dd></dl>

<dl class="class">
<dt id="egglib.tools.GeneticCodes">
<em class="property">class </em><tt class="descclassname">egglib.tools.</tt><tt class="descname">GeneticCodes</tt><a class="headerlink" href="#egglib.tools.GeneticCodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds genetic code. Instantiating this class is pointless since its
contains only class methods.</p>
<dl class="classmethod">
<dt id="egglib.tools.GeneticCodes.codes">
<em class="property">classmethod </em><tt class="descname">codes</tt><big>(</big><big>)</big><a class="headerlink" href="#egglib.tools.GeneticCodes.codes" title="Permalink to this definition">¶</a></dt>
<dd><p>Gives the list of code identifiers. Each code is represented by
three identifiers: <tt class="docutils literal"><span class="pre">(index,</span> <span class="pre">short,</span> <span class="pre">long)</span></tt> where <em>index</em> is the
integer identifier matching NCBI nomenclature (beware that
indices are not consecutive); <em>short</em> is a egglib-defined word
summarizing the code which can be used as an alternative access
means; and <em>long</em> is the full name of the genetic code.</p>
</dd></dl>

<dl class="classmethod">
<dt id="egglib.tools.GeneticCodes.index">
<em class="property">classmethod </em><tt class="descname">index</tt><big>(</big><em>name</em><big>)</big><a class="headerlink" href="#egglib.tools.GeneticCodes.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Tries to identify the index of the genetic code from its short
or full name. Returns <tt class="docutils literal"><span class="pre">None</span></tt> if the string matches no model.
The comparison is case-independent.</p>
</dd></dl>

<dl class="classmethod">
<dt id="egglib.tools.GeneticCodes.is_start">
<em class="property">classmethod </em><tt class="descname">is_start</tt><big>(</big><em>codon</em>, <em>code=1</em><big>)</big><a class="headerlink" href="#egglib.tools.GeneticCodes.is_start" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <tt class="docutils literal"><span class="pre">True</span></tt> if the codon is encoding one the observed
translational start for this genetic code, <tt class="docutils literal"><span class="pre">False</span></tt> otherwise
(including if the codon is invalid). Arguments are similar as
for <a class="reference internal" href="#egglib.tools.translate" title="egglib.tools.translate"><tt class="xref py py-meth docutils literal"><span class="pre">translate()</span></tt></a>.</p>
</dd></dl>

<dl class="classmethod">
<dt id="egglib.tools.GeneticCodes.translate">
<em class="property">classmethod </em><tt class="descname">translate</tt><big>(</big><em>codon</em>, <em>code=1</em><big>)</big><a class="headerlink" href="#egglib.tools.GeneticCodes.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate the codon <em>codon</em> using the indicated code. <em>code</em>
is an identifier (index, short or long name) matching NCBI
nomenclature. Returns the one-letter amino acid code
corresponding to codon, <tt class="docutils literal"><span class="pre">*</span></tt> for stop codon and <tt class="docutils literal"><span class="pre">X</span></tt> for any
invalid codon (string with a length different than 3 or
containing missing data or gaps). The codon specification is
case-independent. Ambiguous codons might still be translated if
the implied possibilities translate all to the same codon. The
IUPAC nomenclature is used. Note that <tt class="docutils literal"><span class="pre">N</span></tt> means <tt class="docutils literal"><span class="pre">A</span></tt>, <tt class="docutils literal"><span class="pre">C</span></tt>,
<tt class="docutils literal"><span class="pre">G</span></tt> or <tt class="docutils literal"><span class="pre">T</span></tt> but that codons containing <tt class="docutils literal"><span class="pre">?</span></tt> or <tt class="docutils literal"><span class="pre">-</span></tt> will
always be translated as <tt class="docutils literal"><span class="pre">X</span></tt>. However, <tt class="docutils literal"><span class="pre">---</span></tt> will be
translated as <tt class="docutils literal"><span class="pre">-</span></tt>.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequence-comparison">
<h2>Sequence comparison<a class="headerlink" href="#sequence-comparison" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="egglib.tools.compare">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">compare</tt><big>(</big><em>seq1</em>, <em>seq2</em><big>)</big><a class="headerlink" href="#egglib.tools.compare" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares two sequences. Sequences are different if they have
different lengths or if they differ by at least one position. The
comparison supports IUPAC ambiguity characters (for example, A and
M are not considered to be different). Furthermore, partially
overlapping ambiguity characters (for example, M and R) are
not taken as different. Returns <tt class="docutils literal"><span class="pre">True</span></tt> is sequences are
identical (or differ only by overlapping IUPAC characters), <tt class="docutils literal"><span class="pre">False</span></tt>
otherwise.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.motifs">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">motifs</tt><big>(</big><em>sequence</em>, <em>motif</em>, <em>mismatches=0</em>, <em>reverse=True</em><big>)</big><a class="headerlink" href="#egglib.tools.motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>Locates motifs in a nucleotide sequence. Standard ambiguity
characters are supported (as explained in <a class="reference internal" href="#egglib.tools.compare" title="egglib.tools.compare"><tt class="xref py py-func docutils literal"><span class="pre">compare()</span></tt></a>
documentation). <em>sequence</em> and <em>motif</em> are nucleotide sequence
strings. <em>mismatches</em> gives the number of nucleotide differences
allowed for motif match. If <em>reverse</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, both strands are
examined (otherwise, only the forward strand is considered. Returns
a list of hits. Each hit is represented by a dictionary containing
keys: <em>start</em>: starting position of the hit, <em>sequence</em>: sequence of
the matching region, <em>mismatches</em>: number of mismatches in the hit,
<em>reverse</em>: <tt class="docutils literal"><span class="pre">True</span></tt> if the hit is on the reverse strand. The hit
position and the found motif are always given with respect to the
passed sequence, even when the motif was found on the reverse hit.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.locate">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">locate</tt><big>(</big><em>sequence</em>, <em>motif</em>, <em>start=0</em>, <em>stop=-1</em><big>)</big><a class="headerlink" href="#egglib.tools.locate" title="Permalink to this definition">¶</a></dt>
<dd><p>Locates the position of the <em>motif</em> in <em>sequence</em>. <em>motif</em> and
<em>sequence</em> should be DNA sequences Ambiguity characters (<tt class="docutils literal"><span class="pre">M</span></tt>,
<tt class="docutils literal"><span class="pre">R</span></tt>, <tt class="docutils literal"><span class="pre">W</span></tt>, <tt class="docutils literal"><span class="pre">S</span></tt>, <tt class="docutils literal"><span class="pre">Y</span></tt>, <tt class="docutils literal"><span class="pre">K</span></tt>, <tt class="docutils literal"><span class="pre">B</span></tt>, <tt class="docutils literal"><span class="pre">D</span></tt>, <tt class="docutils literal"><span class="pre">H</span></tt>, <tt class="docutils literal"><span class="pre">V</span></tt> and
<tt class="docutils literal"><span class="pre">N</span></tt>) are recognized and match the appropriate characters. <tt class="docutils literal"><span class="pre">?</span></tt>
matches any character. Note that the meaning of <tt class="docutils literal"><span class="pre">N</span></tt> (<tt class="docutils literal"><span class="pre">A</span></tt>, <tt class="docutils literal"><span class="pre">C</span></tt>,
<tt class="docutils literal"><span class="pre">G</span></tt> or <tt class="docutils literal"><span class="pre">T</span></tt>) is very different to <tt class="docutils literal"><span class="pre">?</span></tt> (any character). 
<em>start</em> and <em>stop</em> allow to restrict search to a given subset of
<em>sequence</em> (the returned index is still given with respect to the
full sequence). The function returns the position of the first
exact match or, if there is not exact match, the position of the
first matching position allowing ambiguity charactor, or, if there
is no match at all, <tt class="docutils literal"><span class="pre">None</span></tt>.</p>
<p>if <em>hasAmb</em> is <tt class="docutils literal"><span class="pre">True</span></tt>, ambiguities will be supported in the target
sequence (<em>sequence</em>). With that mode on, ambiguities of the motif
sequence (<em>motif</em>) will only be considered as a match if the target
sequence account for all</p>
<div class="versionchanged">
<p><span class="versionmodified">Changed in version 2.1.0: </span>Supports ambiguity characters in <em>sequence</em>. Returns exact
matches first.</p>
</div>
</dd></dl>

</div>
<div class="section" id="tools-not-directly-related-to-sequence-data">
<h2>Tools not (directly) related to sequence data<a class="headerlink" href="#tools-not-directly-related-to-sequence-data" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="egglib.tools.chisquare">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">chisquare</tt><big>(</big><em>ddl</em><big>)</big><a class="headerlink" href="#egglib.tools.chisquare" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the 5% critical value of the chi-square distribution
with <em>ddl</em> degrees of liberty (maximum: 100).</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.correl">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">correl</tt><big>(</big><em>x</em>, <em>y</em><big>)</big><a class="headerlink" href="#egglib.tools.correl" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes correlation coefficients. <em>x</em> is a sequence giving the
values of the explanatory variable. <em>y</em> is a sequence giving the
values of the response variable. Returns a tuple <tt class="docutils literal"><span class="pre">(r,</span> <span class="pre">r**2,</span> <span class="pre">a)</span></tt>
where <em>r</em> is the correlation coefficient and <em>a</em> the regression
coefficient.</p>
</dd></dl>

<dl class="function">
<dt id="egglib.tools.ranges">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">ranges</tt><big>(</big><em>values</em><big>)</big><a class="headerlink" href="#egglib.tools.ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies continuous ranges among the iterable <em>values</em>. <em>values</em>
must be iterable but needs not to be sorted and can contain
duplicates. The function returns a list of <tt class="docutils literal"><span class="pre">(start,stop)</span></tt> tuples,
where <em>start</em> and <em>stop</em> defines a continuous range.</p>
</dd></dl>

<dl class="class">
<dt id="egglib.tools.ReadingFrame">
<em class="property">class </em><tt class="descclassname">egglib.tools.</tt><tt class="descname">ReadingFrame</tt><big>(</big><em>frame</em><big>)</big><a class="headerlink" href="#egglib.tools.ReadingFrame" title="Permalink to this definition">¶</a></dt>
<dd><p>Handles reading frame positions.</p>
<p><em>frame</em> must be a sequence of <tt class="docutils literal"><span class="pre">(start,</span> <span class="pre">stop,</span> <span class="pre">codon_start)</span></tt>
sequences where <em>start</em> and <em>stop</em> gives the first and last
position of an exon and  <em>codon_start</em> is 1 if the first position
of the exon is the first position of a codon (e.g. <tt class="docutils literal"><span class="pre">ATG</span> <span class="pre">ATG</span></tt>),
2 if the first position of the segment is the second position of
a codon (e.g. <tt class="docutils literal"><span class="pre">TG</span> <span class="pre">ATG</span></tt>), 3 if the first position of the
segment is the third position a of codon (e.g. <tt class="docutils literal"><span class="pre">G</span> <span class="pre">ATG</span></tt>), or
<tt class="docutils literal"><span class="pre">None</span></tt> if the reading frame is continuing the previous exon.
If <em>codon_start</em> of the first segment is <tt class="docutils literal"><span class="pre">None</span></tt>, 1 will be
assumed. It is not possible to modify the codon positions held
by the instance after construction.</p>
<dl class="method">
<dt id="egglib.tools.ReadingFrame.codon">
<tt class="descname">codon</tt><big>(</big><em>x</em><big>)</big><a class="headerlink" href="#egglib.tools.ReadingFrame.codon" title="Permalink to this definition">¶</a></dt>
<dd><p>If the position <tt class="docutils literal"><span class="pre">x</span></tt> falls in a complete codon, returns the
three positions of that codon. If <tt class="docutils literal"><span class="pre">x</span></tt> fall outside of defined
segments, or in a codon that appears not to be completely
available, returns <tt class="docutils literal"><span class="pre">None</span></tt>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">the codon positions are cached at build time. As a result,
the result of this method will be incorrect if frame
positions are changed after the creation of the instance.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.ReadingFrame.codons">
<tt class="descname">codons</tt><big>(</big><big>)</big><a class="headerlink" href="#egglib.tools.ReadingFrame.codons" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of complete codons (as triplets of absolute
positions).</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.ReadingFrame.exon">
<tt class="descname">exon</tt><big>(</big><em>x</em><big>)</big><a class="headerlink" href="#egglib.tools.ReadingFrame.exon" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the exon index of a position. Returns <tt class="docutils literal"><span class="pre">-1</span></tt> if the
position falls outside specified segments (out of ranges or
in introns).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="egglib.tools.Updater">
<em class="property">class </em><tt class="descclassname">egglib.tools.</tt><tt class="descname">Updater</tt><big>(</big><em>target=None</em><big>)</big><a class="headerlink" href="#egglib.tools.Updater" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper designed to monitor progress of long-running tasks. In
principle, <a class="reference internal" href="#egglib.tools.Updater" title="egglib.tools.Updater"><tt class="xref py py-class docutils literal"><span class="pre">Updater</span></tt></a> should be coupled to a repetitive
process with a fixed and known number of steps to perform (<em>target</em>)
each requiring the same amount of time. <a class="reference internal" href="#egglib.tools.Updater" title="egglib.tools.Updater"><tt class="xref py py-class docutils literal"><span class="pre">Updater</span></tt></a> should be
updated regularly at reasonnable intervals (not to short to keep it
from being itself a resource load).</p>
<p>The class can be used as in the following examples:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">egglib</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">updater</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">Updater</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1000</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1"># ... time-consuming task here ...</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">updater</span><span class="o">.</span><span class="n">refresh</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">updater</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># The number of iterations is not known - custom display</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">updater</span> <span class="o">=</span> <span class="n">egglib</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">Updater</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">maxi</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">X</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">if</span> <span class="n">X</span><span class="o">&gt;</span><span class="n">maxi</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="n">maxi</span> <span class="o">=</span> <span class="n">X</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">updater</span><span class="o">.</span><span class="n">refresh</span><span class="p">(</span><span class="s1">&#39;$ELAPSED, max. value: </span><span class="si">%f</span><span class="s1">&#39;</span> <span class="o">%</span><span class="n">maxi</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">if</span> <span class="n">X</span><span class="o">&gt;</span><span class="mf">0.99999</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="k">break</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">updater</span><span class="o">.</span><span class="n">refresh</span><span class="p">(</span><span class="s1">&#39;$DONE tries, time $ELAPSED, got </span><span class="si">%f</span><span class="s1">&#39;</span> <span class="o">%</span><span class="n">maxi</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">updater</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Constructor&#8217;s argument <em>target</em> gives the number of iterations
to perform. If <tt class="docutils literal"><span class="pre">None</span></tt>, this information is not available.</p>
<dl class="method">
<dt id="egglib.tools.Updater.close">
<tt class="descname">close</tt><big>(</big><big>)</big><a class="headerlink" href="#egglib.tools.Updater.close" title="Permalink to this definition">¶</a></dt>
<dd><p>If anything was written using <a class="reference internal" href="#egglib.tools.Updater.refresh" title="egglib.tools.Updater.refresh"><tt class="xref py py-meth docutils literal"><span class="pre">refresh()</span></tt></a>, writes any cached
refresh data and writes a new line. Otherwise, does nothing.
This method is automatically called upon object destruction.</p>
</dd></dl>

<dl class="attribute">
<dt id="egglib.tools.Updater.closed">
<tt class="descname">closed</tt><a class="headerlink" href="#egglib.tools.Updater.closed" title="Permalink to this definition">¶</a></dt>
<dd><p><tt class="docutils literal"><span class="pre">True</span></tt> if the instance has been     closed using the <a class="reference internal" href="#egglib.tools.Updater.close" title="egglib.tools.Updater.close"><tt class="xref py py-meth docutils literal"><span class="pre">close()</span></tt></a> method.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.Updater.format">
<tt class="descname">format</tt><big>(</big><em>template=None</em>, <em>increment=1</em><big>)</big><a class="headerlink" href="#egglib.tools.Updater.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string providing feedback about the run&#8217;s progress. 
<em>template</em> gives the template of the string to return. Actual
values will substitute to the following special strings:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">$DONE</span></tt>: Number of steps done.</li>
<li><tt class="docutils literal"><span class="pre">$TARGET</span></tt>: Total number of steps to do.</li>
<li><tt class="docutils literal"><span class="pre">$TODO</span></tt>: Number of steps left to do.</li>
<li><tt class="docutils literal"><span class="pre">$ELAPSED</span></tt>: Time used since object creation.</li>
<li><tt class="docutils literal"><span class="pre">$REMAINING</span></tt>: Estimated time to complete the task.</li>
<li><tt class="docutils literal"><span class="pre">$LREMAINING</span></tt>: Like <tt class="docutils literal"><span class="pre">$REMAINING</span></tt> but computed from the last time point.</li>
<li><tt class="docutils literal"><span class="pre">$TOTAL</span></tt>: Estimated total time (computed as <tt class="docutils literal"><span class="pre">$REMAINING</span></tt> + <tt class="docutils literal"><span class="pre">$ELAPSED</span></tt>)</li>
<li><tt class="docutils literal"><span class="pre">$PERCENT</span></tt>: Percentage done (including <tt class="docutils literal"><span class="pre">%</span></tt> symbol).</li>
</ul>
</div></blockquote>
<p>If <em>template</em> is <tt class="docutils literal"><span class="pre">None</span></tt>, a template defined at construction
time is used. This template is <tt class="docutils literal"><span class="pre">$DONE|$ELAPSED</span></tt> when <em>target</em>
is <tt class="docutils literal"><span class="pre">None</span></tt> and <tt class="docutils literal"><span class="pre">$DONE/$TARGET</span> <span class="pre">(remaining:</span> <span class="pre">$REMAINING)</span></tt> if <em>target</em>
is specified. It is stored at the object attribute
<tt class="xref py py-attr docutils literal"><span class="pre">template</span></tt> and can be modified dynamically.</p>
<p>If <em>increment</em> is different than zero, <a class="reference internal" href="#egglib.tools.Updater.increment" title="egglib.tools.Updater.increment"><tt class="xref py py-meth docutils literal"><span class="pre">increment()</span></tt></a> is
called and this number is passed before formatting the string.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.Updater.increment">
<tt class="descname">increment</tt><big>(</big><em>number</em><big>)</big><a class="headerlink" href="#egglib.tools.Updater.increment" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds <em>number</em> steps and update elapsed time and estimated
running (if <em>target</em> was given). <em>number</em> might be negative.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.Updater.refresh">
<tt class="descname">refresh</tt><big>(</big><em>template=None</em>, <em>increment=1</em>, <em>grain=1.0</em><big>)</big><a class="headerlink" href="#egglib.tools.Updater.refresh" title="Permalink to this definition">¶</a></dt>
<dd><p>This method generates a string exactly as <a class="reference internal" href="#egglib.tools.Updater.format" title="egglib.tools.Updater.format"><tt class="xref py py-meth docutils literal"><span class="pre">format()</span></tt></a> does,
but writes the string to <tt class="xref py py-class docutils literal"><span class="pre">sys.stdout</span></tt> instead. If the
same object has already wrote anything, an equivalent number
of backspaces are written, in principle allowing to overwrite
the previous string and making the string appear to update
itself. The result might no be so nice if something else is
written to the console in the mean time or if the console
doesn&#8217;t support backspaces. This method doesn&#8217;t write a newline,
but the object will upon destruction or call to <a class="reference internal" href="#egglib.tools.Updater.close" title="egglib.tools.Updater.close"><tt class="xref py py-meth docutils literal"><span class="pre">close()</span></tt></a>.
The string will be stripped is it is longer than the object
attribute <tt class="xref py py-attr docutils literal"><span class="pre">length_max</span></tt>, which can be changed dynamically.
If less than the number given by <em>grain</em> (in seconds) has
occurred since the last refresh, nothing is printed.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.Updater.stats">
<tt class="descname">stats</tt><big>(</big><big>)</big><a class="headerlink" href="#egglib.tools.Updater.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary with the current values of counters.</p>
</dd></dl>

<dl class="method">
<dt id="egglib.tools.Updater.wipe">
<tt class="descname">wipe</tt><big>(</big><big>)</big><a class="headerlink" href="#egglib.tools.Updater.wipe" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes an empty line of the maximal possible length, therefore
clearing the line from characters printed by another process
(provided that these characters are not too many).</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="egglib.tools.wrap">
<tt class="descclassname">egglib.tools.</tt><tt class="descname">wrap</tt><big>(</big><em>string</em>, <em>length</em>, <em>indent=0</em><big>)</big><a class="headerlink" href="#egglib.tools.wrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats the string <em>string</em>  to ensures the line lengths are not
larger than <em>length</em>. The optional argument <em>indent</em> specifies the
number of spaces to insert at the beginning of all lines except the
first. The line breaks are inserted at spaces.</p>
<dl class="docutils">
<dt>An example is given below::</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">egglib</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">string</span> <span class="o">=</span> <span class="s2">&quot;Lekrrjf djdhs eeir djs ehehf bnreh eurvz rhffdvfu dksgta.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">egglib</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">wrap</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="go">Lekrrjf djdhs eeir</span>
<span class="go">    djs ehehf bnreh</span>
<span class="go">    eurvz rhffdvfu</span>
<span class="go">    dksgta.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="wrappers.html" title="wrappers"
             >next</a> |</li>
        <li class="right" >
          <a href="data.html" title="Top-level Egglib components"
             >previous</a> |</li>
        <li><a href="index.html">EggLib</a> &raquo;</li>
          <li><a href="egglib-py.html" >Python module</a> &raquo;</li> 
      </ul>
    </div>
        <address style="text-align: right;"><small>Hosted by&nbsp;
        <a href="http://sourceforge.net/projects/seqlib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get seqlib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
        </address>

  </body>
</html>