<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>EggLib C++ library reference manual</title><link href="doxygen.css" rel="stylesheet" type="text/css">
<head>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname"><a href="../index.html">Egglib</a>&#160;<span id="projectnumber">2.1.11</span></div>
   <div id="projectbrief">C++&#160;library&#160;reference&#160;manual</div>
  </td>
 </tr>
 </tbody>
</table>
</head>
</html>
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>egglib</b></li><li class="navelem"><a class="el" href="classegglib_1_1HaplotypeDiversity.html">HaplotypeDiversity</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classegglib_1_1HaplotypeDiversity-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">HaplotypeDiversity Class Reference<div class="ingroups"><a class="el" href="group__polymorphism.html">polymorphism</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Computes diversity based on haplotype analysis.  
 <a href="classegglib_1_1HaplotypeDiversity.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="HaplotypeDiversity_8hpp_source.html">HaplotypeDiversity.hpp</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for HaplotypeDiversity:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classegglib_1_1HaplotypeDiversity.png" usemap="#HaplotypeDiversity_map" alt=""/>
  <map id="HaplotypeDiversity_map" name="HaplotypeDiversity_map">
<area href="classegglib_1_1BaseDiversity.html" title="Base class of diversity classes. " alt="BaseDiversity" shape="rect" coords="0,0,119,24"/>
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a99ef029ae9d7a6e580bc33dab482d2a7"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a99ef029ae9d7a6e580bc33dab482d2a7">HaplotypeDiversity</a> ()</td></tr>
<tr class="memdesc:a99ef029ae9d7a6e580bc33dab482d2a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor.  <a href="#a99ef029ae9d7a6e580bc33dab482d2a7">More...</a><br /></td></tr>
<tr class="separator:a99ef029ae9d7a6e580bc33dab482d2a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9a650084c23cc243450f98a6c80c56d9"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a9a650084c23cc243450f98a6c80c56d9">~HaplotypeDiversity</a> ()</td></tr>
<tr class="memdesc:a9a650084c23cc243450f98a6c80c56d9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor.  <a href="#a9a650084c23cc243450f98a6c80c56d9">More...</a><br /></td></tr>
<tr class="separator:a9a650084c23cc243450f98a6c80c56d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc65520af48c6b6330b79b8f18c55029"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#afc65520af48c6b6330b79b8f18c55029">load</a> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, bool allowMultipleMutations=false, unsigned int ignoreFrequency=0, std::string characterMapping=<a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a>)</td></tr>
<tr class="memdesc:afc65520af48c6b6330b79b8f18c55029"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identifies polymorphic sites and computes basis statistics.  <a href="#afc65520af48c6b6330b79b8f18c55029">More...</a><br /></td></tr>
<tr class="separator:afc65520af48c6b6330b79b8f18c55029"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5bfd555b92f5c9c9e789d1274cffaa81"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5bfd555b92f5c9c9e789d1274cffaa81"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a5bfd555b92f5c9c9e789d1274cffaa81">K</a> () const </td></tr>
<tr class="memdesc:a5bfd555b92f5c9c9e789d1274cffaa81"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of distinct haplotypes. <br /></td></tr>
<tr class="separator:a5bfd555b92f5c9c9e789d1274cffaa81"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a192ec866db4a5e8454cfed9888d5fab4"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a192ec866db4a5e8454cfed9888d5fab4"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a192ec866db4a5e8454cfed9888d5fab4">He</a> () const </td></tr>
<tr class="memdesc:a192ec866db4a5e8454cfed9888d5fab4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Haplotype diversity (unbiased) <br /></td></tr>
<tr class="separator:a192ec866db4a5e8454cfed9888d5fab4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a25751a798a7e30457def89888e1d1359"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a25751a798a7e30457def89888e1d1359">haplotypeIndex</a> (unsigned int) const </td></tr>
<tr class="memdesc:a25751a798a7e30457def89888e1d1359"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the allele number of a given sequence.  <a href="#a25751a798a7e30457def89888e1d1359">More...</a><br /></td></tr>
<tr class="separator:a25751a798a7e30457def89888e1d1359"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a393cdba28596385621a793d5936d7e51"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a393cdba28596385621a793d5936d7e51"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a393cdba28596385621a793d5936d7e51">Kst</a> () const </td></tr>
<tr class="memdesc:a393cdba28596385621a793d5936d7e51"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classegglib_1_1Population.html" title="Handles a single population. ">Population</a> differenciation, based on nucleotides (Hudson 1992a) <br /></td></tr>
<tr class="separator:a393cdba28596385621a793d5936d7e51"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae6e2fbaf8a90e406b727684eb9d3e775"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae6e2fbaf8a90e406b727684eb9d3e775"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#ae6e2fbaf8a90e406b727684eb9d3e775">Fst</a> () const </td></tr>
<tr class="memdesc:ae6e2fbaf8a90e406b727684eb9d3e775"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classegglib_1_1Population.html" title="Handles a single population. ">Population</a> differenciation, based on nucleotides (Hudson 1992b) <br /></td></tr>
<tr class="separator:ae6e2fbaf8a90e406b727684eb9d3e775"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58526d43585fa70d345e4350afcfc194"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a58526d43585fa70d345e4350afcfc194"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a58526d43585fa70d345e4350afcfc194">Gst</a> () const </td></tr>
<tr class="memdesc:a58526d43585fa70d345e4350afcfc194"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classegglib_1_1Population.html" title="Handles a single population. ">Population</a> differenciation, based on haplotypes (Nei version) <br /></td></tr>
<tr class="separator:a58526d43585fa70d345e4350afcfc194"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1bf3c54d1610403ce5ed95a3763df37f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1bf3c54d1610403ce5ed95a3763df37f"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#a1bf3c54d1610403ce5ed95a3763df37f">Hst</a> () const </td></tr>
<tr class="memdesc:a1bf3c54d1610403ce5ed95a3763df37f"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classegglib_1_1Population.html" title="Handles a single population. ">Population</a> differenciation, based on haplotypes (Hudson et al. version) <br /></td></tr>
<tr class="separator:a1bf3c54d1610403ce5ed95a3763df37f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1e1c32ece6640e3d5ba6fe95367f00b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab1e1c32ece6640e3d5ba6fe95367f00b"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1HaplotypeDiversity.html#ab1e1c32ece6640e3d5ba6fe95367f00b">Snn</a> () const </td></tr>
<tr class="memdesc:ab1e1c32ece6640e3d5ba6fe95367f00b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Hudson's Snn (nearest neighbor statistics) <br /></td></tr>
<tr class="separator:ab1e1c32ece6640e3d5ba6fe95367f00b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a728098ee7906f24ec76a07f88d848a65">BaseDiversity</a> ()</td></tr>
<tr class="memdesc:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor.  <a href="#a728098ee7906f24ec76a07f88d848a65">More...</a><br /></td></tr>
<tr class="separator:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a2155869a8e19b5d53b6d3f4eabb9de42">~BaseDiversity</a> ()</td></tr>
<tr class="memdesc:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor.  <a href="#a2155869a8e19b5d53b6d3f4eabb9de42">More...</a><br /></td></tr>
<tr class="separator:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#af9121ee3a3a8722b2d36961a802e8189">reserve</a> (unsigned int numberOfSites)</td></tr>
<tr class="memdesc:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reserve sufficient memory for a given number of polymorphic sites.  <a href="#af9121ee3a3a8722b2d36961a802e8189">More...</a><br /></td></tr>
<tr class="separator:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a77fd109923331c4c39106ca1b99d5e57"></a>
const <a class="el" href="classegglib_1_1SitePolymorphism.html">SitePolymorphism</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a77fd109923331c4c39106ca1b99d5e57">get_site</a> (unsigned int index) const </td></tr>
<tr class="memdesc:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a site. <br /></td></tr>
<tr class="separator:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2d827860e77ed18cb0356640561abf2f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a2d827860e77ed18cb0356640561abf2f">get_position</a> (unsigned int index) const </td></tr>
<tr class="memdesc:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a site position. <br /></td></tr>
<tr class="separator:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad20897c5c8bd47f5d4005989bead0e55"></a>
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#ad20897c5c8bd47f5d4005989bead0e55">reset</a> ()</td></tr>
<tr class="memdesc:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears and re-initializes object. <br /></td></tr>
<tr class="separator:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a02fd73d861ef2e4aabb38c0c9ff82947"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a02fd73d861ef2e4aabb38c0c9ff82947"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>init</b> ()</td></tr>
<tr class="separator:a02fd73d861ef2e4aabb38c0c9ff82947"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8bb3912a3ce86b15842e79d0b421204"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8bb3912a3ce86b15842e79d0b421204"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>clear</b> ()</td></tr>
<tr class="separator:ac8bb3912a3ce86b15842e79d0b421204"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abf60f121d39db164451dc508e3943625"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abf60f121d39db164451dc508e3943625"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>diff</b> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, unsigned int ind1, unsigned int ind2) const </td></tr>
<tr class="separator:abf60f121d39db164451dc508e3943625"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_methods_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pro_methods_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Protected Member Functions inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a9a7c8b0679c8aa67a65808ea073688a9 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9a7c8b0679c8aa67a65808ea073688a9"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>importSites</b> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, bool allowMultipleMutations, double minimumExploitableData, unsigned int ignoreFrequency, std::string characterMapping, bool useZeroAsAncestral, bool ignoreOutgroup)</td></tr>
<tr class="separator:a9a7c8b0679c8aa67a65808ea073688a9 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af43dec4a12ccce2c941edebcccd6c36a inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af43dec4a12ccce2c941edebcccd6c36a"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>analyzeSite</b> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, unsigned int index, double maxMissingData, bool ignoreOutgroup)</td></tr>
<tr class="separator:af43dec4a12ccce2c941edebcccd6c36a inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b4928f332a83517fe59a8624136a249 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8b4928f332a83517fe59a8624136a249"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>getPopIndex</b> (unsigned int label) const </td></tr>
<tr class="separator:a8b4928f332a83517fe59a8624136a249 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:a91cfd0ddc0c2346973f9d15493b2b551"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a91cfd0ddc0c2346973f9d15493b2b551"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>m_loaded</b></td></tr>
<tr class="separator:a91cfd0ddc0c2346973f9d15493b2b551"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a22da9cfce8ebcd09195ef0dc962b38f2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a22da9cfce8ebcd09195ef0dc962b38f2"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>m_K</b></td></tr>
<tr class="separator:a22da9cfce8ebcd09195ef0dc962b38f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a41a37d09cc520800fa676db48cc3cd16"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a41a37d09cc520800fa676db48cc3cd16"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_He</b></td></tr>
<tr class="separator:a41a37d09cc520800fa676db48cc3cd16"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad9dc296d7d11f555a8a28b2f3c48f338"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad9dc296d7d11f555a8a28b2f3c48f338"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_Kst</b></td></tr>
<tr class="separator:ad9dc296d7d11f555a8a28b2f3c48f338"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a355b0f2d3b074e7bb4774007c6235b0b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a355b0f2d3b074e7bb4774007c6235b0b"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_Fst</b></td></tr>
<tr class="separator:a355b0f2d3b074e7bb4774007c6235b0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa301577c8d3cc0fbcfedc4d11e6641ea"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa301577c8d3cc0fbcfedc4d11e6641ea"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_Gst</b></td></tr>
<tr class="separator:aa301577c8d3cc0fbcfedc4d11e6641ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a62310650bf3389611c7efabd1203d98e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a62310650bf3389611c7efabd1203d98e"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_Hst</b></td></tr>
<tr class="separator:a62310650bf3389611c7efabd1203d98e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2298d0a987b55441d51a87c2cb8ab07c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2298d0a987b55441d51a87c2cb8ab07c"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>m_Snn</b></td></tr>
<tr class="separator:a2298d0a987b55441d51a87c2cb8ab07c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2bb110b4e4d467247f6b20aff9107bb2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2bb110b4e4d467247f6b20aff9107bb2"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>m_haplotypeIndex</b></td></tr>
<tr class="separator:a2bb110b4e4d467247f6b20aff9107bb2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_attribs_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pro_attribs_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Protected Attributes inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a2aa7a2be99f12e471f767cee98d6a1f1 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2aa7a2be99f12e471f767cee98d6a1f1"></a>
<a class="el" href="classegglib_1_1SitePolymorphism.html">SitePolymorphism</a> **&#160;</td><td class="memItemRight" valign="bottom"><b>v_sites</b></td></tr>
<tr class="separator:a2aa7a2be99f12e471f767cee98d6a1f1 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a65260e65d24391c727ff74dafd05411a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a65260e65d24391c727ff74dafd05411a"></a>
bool *&#160;</td><td class="memItemRight" valign="bottom"><b>v_orientables</b></td></tr>
<tr class="separator:a65260e65d24391c727ff74dafd05411a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f24b85598b7bfbd92e59a2aa816468a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0f24b85598b7bfbd92e59a2aa816468a"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_sitePositions</b></td></tr>
<tr class="separator:a0f24b85598b7bfbd92e59a2aa816468a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acaa4bf5869424fc58e4f454ad32983ef inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acaa4bf5869424fc58e4f454ad32983ef"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_reserved</b></td></tr>
<tr class="separator:acaa4bf5869424fc58e4f454ad32983ef inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a49f7b76263e8d660cad6bab0b1a0ee18 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a49f7b76263e8d660cad6bab0b1a0ee18"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_ns</b></td></tr>
<tr class="separator:a49f7b76263e8d660cad6bab0b1a0ee18 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1d5cf6776213289afae59d1d5013ef3 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab1d5cf6776213289afae59d1d5013ef3"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_S</b></td></tr>
<tr class="separator:ab1d5cf6776213289afae59d1d5013ef3 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a951aa44c6e7034205b92087a5b0ad1be inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a951aa44c6e7034205b92087a5b0ad1be"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_So</b></td></tr>
<tr class="separator:a951aa44c6e7034205b92087a5b0ad1be inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ad85c1898fbeba3f7a795a5d8901812 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6ad85c1898fbeba3f7a795a5d8901812"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_eta</b></td></tr>
<tr class="separator:a6ad85c1898fbeba3f7a795a5d8901812 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa3ed42221d826288582cd5d4ffe561b2 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa3ed42221d826288582cd5d4ffe561b2"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_nseff</b></td></tr>
<tr class="separator:aa3ed42221d826288582cd5d4ffe561b2 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb38ffa03ab28fcdf786dbad7cf3c800 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aeb38ffa03ab28fcdf786dbad7cf3c800"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_lseff</b></td></tr>
<tr class="separator:aeb38ffa03ab28fcdf786dbad7cf3c800 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f748a5c62571140a7a3df61711190f4 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7f748a5c62571140a7a3df61711190f4"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_nseffo</b></td></tr>
<tr class="separator:a7f748a5c62571140a7a3df61711190f4 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c42f196ad1fb4cdf0031bb4acc53d6f inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3c42f196ad1fb4cdf0031bb4acc53d6f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_lseffo</b></td></tr>
<tr class="separator:a3c42f196ad1fb4cdf0031bb4acc53d6f inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3f8ba820cd11d01df127e058ab4710b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac3f8ba820cd11d01df127e058ab4710b"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_npop</b></td></tr>
<tr class="separator:ac3f8ba820cd11d01df127e058ab4710b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8b67942007a1d99348372b3fd25e65b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8b67942007a1d99348372b3fd25e65b"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_popLabel</b></td></tr>
<tr class="separator:ac8b67942007a1d99348372b3fd25e65b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac93cdae4eb273f9e3641d94eb90d93c9 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac93cdae4eb273f9e3641d94eb90d93c9"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>p_allowMultipleMutations</b></td></tr>
<tr class="separator:ac93cdae4eb273f9e3641d94eb90d93c9 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a971874b4d9cb1c5737ec7e1d73c4a6f5 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a971874b4d9cb1c5737ec7e1d73c4a6f5"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>p_minimumExploitableData</b></td></tr>
<tr class="separator:a971874b4d9cb1c5737ec7e1d73c4a6f5 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb27bce76989a6521804ba5f618efe87 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb27bce76989a6521804ba5f618efe87"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><b>p_characterMapping</b></td></tr>
<tr class="separator:afb27bce76989a6521804ba5f618efe87 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a12a9ec49e72be17b7c4993df7ff20903 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a12a9ec49e72be17b7c4993df7ff20903"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>p_pos_sep_mapping</b></td></tr>
<tr class="separator:a12a9ec49e72be17b7c4993df7ff20903 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58c7dc12609f51192588cbc5b08c1f5b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a58c7dc12609f51192588cbc5b08c1f5b"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>p_useZeroAsAncestral</b></td></tr>
<tr class="separator:a58c7dc12609f51192588cbc5b08c1f5b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acdeb3544a8b488558119e39ed5b27035 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acdeb3544a8b488558119e39ed5b27035"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>p_ignoreFrequency</b></td></tr>
<tr class="separator:acdeb3544a8b488558119e39ed5b27035 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="inherited"></a>
Additional Inherited Members</h2></td></tr>
<tr class="inherit_header pub_static_attribs_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pub_static_attribs_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Static Public Attributes inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a> = &quot;ACGT MRWSYKBDHVN?-&quot;</td></tr>
<tr class="memdesc:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for DNA data.  <a href="#a1e968823aeb9f9c279596cc3c2b93431">More...</a><br /></td></tr>
<tr class="separator:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a481307ebf500b6b714e767694019be94">rnaMapping</a> = &quot;ACGU MRWSYKBDHVN?-&quot;</td></tr>
<tr class="memdesc:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for RNA data.  <a href="#a481307ebf500b6b714e767694019be94">More...</a><br /></td></tr>
<tr class="separator:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#aa93366ce287a12c730120fc0e1cd2875">aaMapping</a> = &quot;GALMFWKQESPVICYHRNDT X-&quot;</td></tr>
<tr class="memdesc:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for amino acid data.  <a href="#aa93366ce287a12c730120fc0e1cd2875">More...</a><br /></td></tr>
<tr class="separator:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Computes diversity based on haplotype analysis. </p>
<p>This class relies on detection of polymorphic sites, as does <a class="el" href="classegglib_1_1NucleotideDiversity.html" title="Performs analyzes of population genetics. ">NucleotideDiversity</a>, with the exception that sites with missing data cannot be processed (minimumExploitableData is enforced to 1.).</p>
<p>Like <a class="el" href="classegglib_1_1NucleotideDiversity.html" title="Performs analyzes of population genetics. ">NucleotideDiversity</a>, the same object can be used to analyze different data sets. Only the call to <a class="el" href="classegglib_1_1HaplotypeDiversity.html#afc65520af48c6b6330b79b8f18c55029" title="Identifies polymorphic sites and computes basis statistics. ">load()</a> is required before accessing the data.</p>
<p>Hst, Gst and Kst are between population differenciation indices. They are respectively defined in equations 2, 5-6 and 9 of Hudson et al. 1992a (Molecular Biology and Evolution 9:138-151). Also, Fst is defined in equation 3 of Hudson et al. 1992b (Genetics 132:583-589). Finally, Snn is from Hudson 2000 Genetics. It is computed as the average of Xi for all sequences. Where Xi is the ratio of nearest neighbours from the same group to the number of nearest neighbours. Nearest neigbours are all the sequences with the lowest number of differences to the focal sequence. NOTE: Gst/Hst are quite similar, but Fst and Kst are more different. Snn is a different statistic. Gst and Hst are two ways to estimate the between-population fraction of haplotypic diversity. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a99ef029ae9d7a6e580bc33dab482d2a7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1HaplotypeDiversity.html">HaplotypeDiversity</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor. </p>

</div>
</div>
<a class="anchor" id="a9a650084c23cc243450f98a6c80c56d9"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">~<a class="el" href="classegglib_1_1HaplotypeDiversity.html">HaplotypeDiversity</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Destructor. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a25751a798a7e30457def89888e1d1359"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int haplotypeIndex </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the allele number of a given sequence. </p>
<p>The passed index must be given ignoring any outgroup sequence. </p>

</div>
</div>
<a class="anchor" id="afc65520af48c6b6330b79b8f18c55029"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void load </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;&#160;</td>
          <td class="paramname"><em>data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allowMultipleMutations</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>ignoreFrequency</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>characterMapping</em> = <code><a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a></code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identifies polymorphic sites and computes basis statistics. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">data</td><td>an alignment object (subclass of <a class="el" href="classegglib_1_1CharMatrix.html" title="Interface for classes usable as a square matrix of characters. ">CharMatrix</a>). The presence of outgroup or of different populations will be detected based on the populationLabel members of the passed object. The populationLabel 999 will be interpreted as outgroups. If several outgroups are passed, sites were the outgroups are not consistent will be treated as "non-
orientable".</td></tr>
    <tr><td class="paramname">allowMultipleMutations</td><td>if true, sites with more than two alleles will not be ignored. The sum of the frequencies of all alleles not matching the outgroup will treated as the derived allele frequency (for orientable sites).</td></tr>
    <tr><td class="paramname">ignoreFrequency</td><td>removes sites that are polymorph because of an allele at absolute frequency smaller than or equal to this value. If ignoreFrequency=1, no sites are removed, if ignoreFrequency=1, singleton sites are ignored. Such sites are completely removed from the analysis (not counted in lseff). Note that if more than one mutation is allowed, the site is removed only if all the alleles but one are smaller than or equal to this value. For example, an alignment column AAAAAAGAAT is ignored with an ignoreFrequency of 1, but AAAAAAGGAT is conserved (including the third allele T which is a singleton).</td></tr>
    <tr><td class="paramname">characterMapping</td><td>a string giving the list of characters that should be considered as valid data. If a space is present in the string, the characters left of the space will be treated as valid data and the characters right of the space will be treated as missing data, that is tolerated but ignored. All characters not in the string will cause an <a class="el" href="classegglib_1_1EggInvalidCharacterError.html" title="Exception type for invalid character. ">EggInvalidCharacterError</a> to be raised. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/<a class="el" href="HaplotypeDiversity_8hpp_source.html">HaplotypeDiversity.hpp</a></li>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/HaplotypeDiversity.cpp</li>
</ul>
</div><!-- contents -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head></head><body><hr size="1">
<address style="text-align: right;"><small>Hosted by&nbsp;</address>
<a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
</body></html>
