<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>EggLib C++ library reference manual</title><link href="doxygen.css" rel="stylesheet" type="text/css">
<head>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname"><a href="../index.html">Egglib</a>&#160;<span id="projectnumber">2.1.11</span></div>
   <div id="projectbrief">C++&#160;library&#160;reference&#160;manual</div>
  </td>
 </tr>
 </tbody>
</table>
</head>
</html>
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>egglib</b></li><li class="navelem"><a class="el" href="classegglib_1_1Convert.html">Convert</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-static-attribs">Static Protected Attributes</a> &#124;
<a href="classegglib_1_1Convert-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Convert Class Reference<div class="ingroups"><a class="el" href="group__core.html">core</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Performs conversion between sequence holder types.  
 <a href="classegglib_1_1Convert.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="Convert_8hpp_source.html">Convert.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a2881b336437e660a756f3b90de5c2502"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="classegglib_1_1Align.html">Align</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#a2881b336437e660a756f3b90de5c2502">align</a> (<a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;dataMatrix, unsigned int length=0, <a class="el" href="classegglib_1_1Random.html">Random</a> *random=NULL, bool randomizePositions=false, bool randomizeNonVaryingStates=false, bool randomizeAlleles=false, bool enforceLength=false, std::string mapping=&quot;ACGT&quot;, char unknown='?', char nonVaryingState='A')</td></tr>
<tr class="memdesc:a2881b336437e660a756f3b90de5c2502"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> to <a class="el" href="classegglib_1_1Align.html" title="Handles a sequence alignment. ">Align</a> conversion.  <a href="#a2881b336437e660a756f3b90de5c2502">More...</a><br /></td></tr>
<tr class="separator:a2881b336437e660a756f3b90de5c2502"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8599a89e274d30d29862ef7fc47ef3e"><td class="memItemLeft" align="right" valign="top">static bpp::AlignedSequenceContainer&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#ac8599a89e274d30d29862ef7fc47ef3e">egglib2bpp</a> (<a class="el" href="classegglib_1_1Align.html">Align</a> &amp;<a class="el" href="classegglib_1_1Convert.html#a2881b336437e660a756f3b90de5c2502">align</a>, unsigned int alphabetID, unsigned int outgroupFlag, unsigned int offset=0)</td></tr>
<tr class="memdesc:ac8599a89e274d30d29862ef7fc47ef3e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Converts an alignment to the equivalent Bio++ type.  <a href="#ac8599a89e274d30d29862ef7fc47ef3e">More...</a><br /></td></tr>
<tr class="separator:ac8599a89e274d30d29862ef7fc47ef3e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a8ec835af4f74f124dab78001882a9ea9"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#a8ec835af4f74f124dab78001882a9ea9">Convert</a> ()</td></tr>
<tr class="memdesc:a8ec835af4f74f124dab78001882a9ea9"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be instantiated.  <a href="#a8ec835af4f74f124dab78001882a9ea9">More...</a><br /></td></tr>
<tr class="separator:a8ec835af4f74f124dab78001882a9ea9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8badb36714a277c8c5cf33062a5b3bc7"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#a8badb36714a277c8c5cf33062a5b3bc7">Convert</a> (const <a class="el" href="classegglib_1_1Convert.html">Convert</a> &amp;source)</td></tr>
<tr class="memdesc:a8badb36714a277c8c5cf33062a5b3bc7"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be instantiated.  <a href="#a8badb36714a277c8c5cf33062a5b3bc7">More...</a><br /></td></tr>
<tr class="separator:a8badb36714a277c8c5cf33062a5b3bc7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab23b8dbcc50b1e9087ccc4fb16bc3103"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classegglib_1_1Convert.html">Convert</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#ab23b8dbcc50b1e9087ccc4fb16bc3103">operator=</a> (const <a class="el" href="classegglib_1_1Convert.html">Convert</a> &amp;source)</td></tr>
<tr class="memdesc:ab23b8dbcc50b1e9087ccc4fb16bc3103"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be instantiated.  <a href="#ab23b8dbcc50b1e9087ccc4fb16bc3103">More...</a><br /></td></tr>
<tr class="separator:ab23b8dbcc50b1e9087ccc4fb16bc3103"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a029bea52cf122d52f37f514e8e6647ee"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Convert.html#a029bea52cf122d52f37f514e8e6647ee">~Convert</a> ()</td></tr>
<tr class="memdesc:a029bea52cf122d52f37f514e8e6647ee"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be instantiated.  <a href="#a029bea52cf122d52f37f514e8e6647ee">More...</a><br /></td></tr>
<tr class="separator:a029bea52cf122d52f37f514e8e6647ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-static-attribs"></a>
Static Protected Attributes</h2></td></tr>
<tr class="memitem:a20769121243e2b39c4e2d64bb63d3bf1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a20769121243e2b39c4e2d64bb63d3bf1"></a>
static bpp::DNA&#160;</td><td class="memItemRight" valign="bottom"><b>dnaAlphabet</b></td></tr>
<tr class="separator:a20769121243e2b39c4e2d64bb63d3bf1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a428ae757fbde2a99608f9847b4784470"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a428ae757fbde2a99608f9847b4784470"></a>
static bpp::RNA&#160;</td><td class="memItemRight" valign="bottom"><b>rnaAlphabet</b></td></tr>
<tr class="separator:a428ae757fbde2a99608f9847b4784470"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44238a8331a4137f3a7f03a5158a0741"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a44238a8331a4137f3a7f03a5158a0741"></a>
static bpp::ProteicAlphabet&#160;</td><td class="memItemRight" valign="bottom"><b>proteicAlphabet</b></td></tr>
<tr class="separator:a44238a8331a4137f3a7f03a5158a0741"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a01f7687758354145a1f685a7c6186351"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a01f7687758354145a1f685a7c6186351"></a>
static bpp::CodonAlphabet&#160;</td><td class="memItemRight" valign="bottom"><b>CodonAlphabet</b></td></tr>
<tr class="separator:a01f7687758354145a1f685a7c6186351"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Performs conversion between sequence holder types. </p>
<p>Static methods of this class allows conversion between sequence holder types implying parametrizable modifications. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a8ec835af4f74f124dab78001882a9ea9"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Convert.html">Convert</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be instantiated. </p>

</div>
</div>
<a class="anchor" id="a8badb36714a277c8c5cf33062a5b3bc7"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Convert.html">Convert</a> </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1Convert.html">Convert</a> &amp;&#160;</td>
          <td class="paramname"><em>source</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be instantiated. </p>

</div>
</div>
<a class="anchor" id="a029bea52cf122d52f37f514e8e6647ee"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">virtual ~<a class="el" href="classegglib_1_1Convert.html">Convert</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be instantiated. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a2881b336437e660a756f3b90de5c2502"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Align.html">Align</a> align </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;&#160;</td>
          <td class="paramname"><em>dataMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>length</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classegglib_1_1Random.html">Random</a> *&#160;</td>
          <td class="paramname"><em>random</em> = <code>NULL</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>randomizePositions</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>randomizeNonVaryingStates</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>randomizeAlleles</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>enforceLength</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>mapping</em> = <code>&quot;ACGT&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>unknown</em> = <code>'?'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>nonVaryingState</em> = <code>'A'</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> to <a class="el" href="classegglib_1_1Align.html" title="Handles a sequence alignment. ">Align</a> conversion. </p>
<p>By defaut, this method generates an <a class="el" href="classegglib_1_1Align.html" title="Handles a sequence alignment. ">Align</a> instance containing only the polymorphic sites. The integers of the <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> will be converted as follow: 0 to A, 1 to C, 2 to G and 3 to T. This behaviour can be largely modified using options.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dataMatrix</td><td><a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> instance.</td></tr>
    <tr><td class="paramname">length</td><td>length of the desired alignment. Non-varying stretches of data will be introduced to reach the specified length. By default the positions of segregating sites will be determined from the positions given by the <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> object. Those positions are expressed in a continuous range, and will be discretized. Mutations falling on the same site will be moved of one position left or right (always preserving the order of mutation sites). If positions are all zero (the default of the <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> class) and if length is larger than the number of segregating sites, then all segregating sites will cluster on the left-hand side of the alignment.</td></tr>
    <tr><td class="paramname">random</td><td>the address to a <a class="el" href="classegglib_1_1Random.html" title="Pseudo-random number generator. ">Random</a> object allowing to draw random numbers (for randomizing positions and/or non-varying states). If an address is provided but no random numbers are required, it is ignored. If no address if provided and random numbers are required, a <a class="el" href="classegglib_1_1Random.html" title="Pseudo-random number generator. ">Random</a> instance is built internally.</td></tr>
    <tr><td class="paramname">randomizePositions</td><td>if true, the positions specified in the <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> objects are ignored and the positions of mutations are drawn randomly along the interval (only if the specified length is larger than the number of segregating sites). If randomizePositions and false and positions are not</td></tr>
    <tr><td class="paramname">enforceLength</td><td>specify whether a <a class="el" href="classegglib_1_1EggRuntimeError.html" title="Exception type for runtime errors. ">EggRuntimeError</a> should be thrown when the number of polymorphic sites is larger than the specified length. If false (the default) and in cases where the specified length is too short to harbor all polymorphic sites, the alignment length will be increased as needed.</td></tr>
    <tr><td class="paramname">randomizeNonVaryingStates</td><td>if true, the stretches of conserved positions (between segregating sites) will be randomly drawn from the current symbol mapping. Otherwise, the symbol given by fixed will be used.</td></tr>
    <tr><td class="paramname">randomizeAlleles</td><td>if true, alleles will be drawn randomly from the mapped characters. Note that if a genotype value is larger than the size of the mapping, it will be replaced by the character given by unknown, without randomization. In other words, with the mapping "ACGT", alleles 0, 1, 2 and 3 will be randomly assigned to these four characters, but larger and negative alleles will be assigned to the unknown character.</td></tr>
    <tr><td class="paramname">mapping</td><td>a string given the character to assign to different character values read from the <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a>. If the read value is 0, the first character of the string will used, the the value is 1, the second character will be used, and so on. If the integer read is out of range (in particular, for any negative value), then the character given by unknown will be used. An empty string will always lead to alignments containing only the character given by unknown. The string "01" is suitable for binary data.</td></tr>
    <tr><td class="paramname">unknown</td><td>the character to use if an integer genotype value is not mapped in the mapping string (that is, if the mapping string is too short).</td></tr>
    <tr><td class="paramname">nonVaryingState</td><td>character to use for conserved stretches of data. It doesn't have to be included in the mapping. If randomizeNonVaryingState is true, this argument is ignored.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The resulting <a class="el" href="classegglib_1_1Align.html" title="Handles a sequence alignment. ">Align</a> object. </dd></dl>

</div>
</div>
<a class="anchor" id="ac8599a89e274d30d29862ef7fc47ef3e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bpp::AlignedSequenceContainer egglib2bpp </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1Align.html">Align</a> &amp;&#160;</td>
          <td class="paramname"><em>align</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>alphabetID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>outgroupFlag</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>offset</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Converts an alignment to the equivalent Bio++ type. </p>
<p>During conversion, name information is lost (arbitrary names are generated in order toprevent duplicate names). The object is attached to an alphabet matching the passed integer. The names are bare rank integers (starting at the value giving by <em>offset</em>).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">align</td><td>the source alignment object.</td></tr>
    <tr><td class="paramname">alphabetID</td><td>an integer indicating which alphabet to use:<ul>
<li>1 for DNA</li>
<li>2 for RNA</li>
<li>3 for proteins</li>
<li>4 for codons</li>
</ul>
Other values will result in an exception.</td></tr>
    <tr><td class="paramname">outgroupFlag</td><td>an integer indicating whether to include outgroup sequences:<ul>
<li>0 use all sequences</li>
<li>1 use only sequences without 999 label (ingroup)</li>
<li>2 use only sequences with 999 label (outgroup)</li>
</ul>
Other values will result in an exception.</td></tr>
    <tr><td class="paramname">offset</td><td>enter an integer to shift the names of the resulting alignment (useful to merge alignment and ensure that names are not duplicated).</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A Bio++ alignment. </dd></dl>

</div>
</div>
<a class="anchor" id="ab23b8dbcc50b1e9087ccc4fb16bc3103"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Convert.html">Convert</a>&amp; operator= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1Convert.html">Convert</a> &amp;&#160;</td>
          <td class="paramname"><em>source</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be instantiated. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/<a class="el" href="Convert_8hpp_source.html">Convert.hpp</a></li>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/Convert.cpp</li>
</ul>
</div><!-- contents -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head></head><body><hr size="1">
<address style="text-align: right;"><small>Hosted by&nbsp;</address>
<a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
</body></html>
