<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>EggLib C++ library reference manual</title><link href="doxygen.css" rel="stylesheet" type="text/css">
<head>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname"><a href="../index.html">Egglib</a>&#160;<span id="projectnumber">2.1.11</span></div>
   <div id="projectbrief">C++&#160;library&#160;reference&#160;manual</div>
  </td>
 </tr>
 </tbody>
</table>
</head>
</html>
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>egglib</b></li><li class="navelem"><a class="el" href="classegglib_1_1Ms.html">Ms</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="classegglib_1_1Ms-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Ms Class Reference<div class="ingroups"><a class="el" href="group__core.html">core</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>ms-like sequence format parser  
 <a href="classegglib_1_1Ms.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="Ms_8hpp_source.html">Ms.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ace989f0bb346fb9a4b5b3355e8095708"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#ace989f0bb346fb9a4b5b3355e8095708">get</a> (std::string, unsigned int ns, bool separated=false)</td></tr>
<tr class="memdesc:ace989f0bb346fb9a4b5b3355e8095708"><td class="mdescLeft">&#160;</td><td class="mdescRight">Imports a sequence alignment.  <a href="#ace989f0bb346fb9a4b5b3355e8095708">More...</a><br /></td></tr>
<tr class="separator:ace989f0bb346fb9a4b5b3355e8095708"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a292bb2e60956535c3be080fa744db5ec"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#a292bb2e60956535c3be080fa744db5ec">get</a> (std::istream &amp;stream, unsigned int ns, bool separated=false)</td></tr>
<tr class="memdesc:a292bb2e60956535c3be080fa744db5ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Imports a sequence alignment.  <a href="#a292bb2e60956535c3be080fa744db5ec">More...</a><br /></td></tr>
<tr class="separator:a292bb2e60956535c3be080fa744db5ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae7ba954a4a67bc407318bc8906debafb"><td class="memItemLeft" align="right" valign="top">static std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#ae7ba954a4a67bc407318bc8906debafb">format</a> (<a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;dataMatrix, bool separated=false)</td></tr>
<tr class="memdesc:ae7ba954a4a67bc407318bc8906debafb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Exports a sequence alignment.  <a href="#ae7ba954a4a67bc407318bc8906debafb">More...</a><br /></td></tr>
<tr class="separator:ae7ba954a4a67bc407318bc8906debafb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae92be9506c9096f30afe01f60f12075a"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#ae92be9506c9096f30afe01f60f12075a">format</a> (std::ostream &amp;stream, <a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;dataMatrix, bool separated=false)</td></tr>
<tr class="memdesc:ae92be9506c9096f30afe01f60f12075a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Exports a sequence alignment.  <a href="#ae92be9506c9096f30afe01f60f12075a">More...</a><br /></td></tr>
<tr class="separator:ae92be9506c9096f30afe01f60f12075a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a466e7f268b9a787a6ba108b29fd76b8d"><td class="memItemLeft" align="right" valign="top">static double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#a466e7f268b9a787a6ba108b29fd76b8d">tMRCA</a> ()</td></tr>
<tr class="memdesc:a466e7f268b9a787a6ba108b29fd76b8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the last tMRCA read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance.  <a href="#a466e7f268b9a787a6ba108b29fd76b8d">More...</a><br /></td></tr>
<tr class="separator:a466e7f268b9a787a6ba108b29fd76b8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a006d5efa47cda07ee4daa9e2d4190f13"><td class="memItemLeft" align="right" valign="top">static double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#a006d5efa47cda07ee4daa9e2d4190f13">prob</a> ()</td></tr>
<tr class="memdesc:a006d5efa47cda07ee4daa9e2d4190f13"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the last "prob" read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance.  <a href="#a006d5efa47cda07ee4daa9e2d4190f13">More...</a><br /></td></tr>
<tr class="separator:a006d5efa47cda07ee4daa9e2d4190f13"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad9932fcd08983abecdec54cb658ba53"><td class="memItemLeft" align="right" valign="top">static std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Ms.html#aad9932fcd08983abecdec54cb658ba53">trees</a> ()</td></tr>
<tr class="memdesc:aad9932fcd08983abecdec54cb658ba53"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the tree string found in the last simulation read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance.  <a href="#aad9932fcd08983abecdec54cb658ba53">More...</a><br /></td></tr>
<tr class="separator:aad9932fcd08983abecdec54cb658ba53"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>ms-like sequence format parser </p>
<p>The class provides parsing (input) and formatting (output) operations in ms format, that is the format used by Richard Hudson's program ms for outputting genotypes and by the associated program samplestat for reading them. Both types of operations are available through static methods using either a string or a stream (which can be a stream to or from a file or a string). In either case, types from the STL are used. Although ms deals only with data coded with 0 and 1, the class <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> offers the possibility of both importing and exporting data coded with by integer. All methods have an option named "separated". If this option is true, the parser or formatter introduces a slight modification of the format: genotypes individual data are separated by a white space ("1 0 1 1" instead of "1011", allowing genotype values larger than 9: "1 0 11 1"). </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="ae7ba954a4a67bc407318bc8906debafb"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::string format </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;&#160;</td>
          <td class="paramname"><em>dataMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>separated</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Exports a sequence alignment. </p>
<p>Internally creates a stringstream, calls the overloaded method and returns the outcome.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">dataMatrix</td><td>the alignment object to write. </td></tr>
    <tr><td class="paramname">separated</td><td>true if a white space separator must be placed between the genotype at each site. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ae92be9506c9096f30afe01f60f12075a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void format </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;&#160;</td>
          <td class="paramname"><em>stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> &amp;&#160;</td>
          <td class="paramname"><em>dataMatrix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>separated</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Exports a sequence alignment. </p>
<p>Writes the formatted string to the stream 'on the fly'. The formatted string is guaranteed to starts with a // line and ends with an empty line. The client is expected to take care of writing any header and add an additional white line between simulations if needed. The method throws a SeqlibRuntimeError if the stream is not writable. The data matrix should contain only data within range 0-9 if separated is false (default) and any positive (&gt;=0) integer if separated is true. Note that output generated with separated=true is never compatible with the original ms format, and that output generated with separator=false is compatible with the original ms format only if all alleles are 0 or 1 (which is not checked by this formatted).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">stream</td><td>the stream (file or string stream) where to write the output. </td></tr>
    <tr><td class="paramname">dataMatrix</td><td>the alignment object to write. </td></tr>
    <tr><td class="paramname">separated</td><td>true if a white space separator must be placed between the genotype at each site. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ace989f0bb346fb9a4b5b3355e8095708"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> get </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>str</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>ns</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>separated</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Imports a sequence alignment. </p>
<p>Creates a istringstream from the string and calls the overloaded method.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">str</td><td>the string to parse. </td></tr>
    <tr><td class="paramname">ns</td><td>the expected number of sequences. </td></tr>
    <tr><td class="paramname">separated</td><td>true if a white space separator is placed between genotype at each site.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A sequence alignment as a data matrix. </dd></dl>

</div>
</div>
<a class="anchor" id="a292bb2e60956535c3be080fa744db5ec"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> get </td>
          <td>(</td>
          <td class="paramtype">std::istream &amp;&#160;</td>
          <td class="paramname"><em>stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>ns</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>separated</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Imports a sequence alignment. </p>
<p>Attemps to generate a <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> object from the stream. Reads only one simulation and throws a SeqlibFormatError exception in case of format error.</p>
<p>Allows any number of white lines before the //, but no other data. Supports  at the end of lines (before the <br />
). Accepted symbols are all integers (0-9).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">stream</td><td>the stream to parse. </td></tr>
    <tr><td class="paramname">ns</td><td>the expected number of sequences. </td></tr>
    <tr><td class="paramname">separated</td><td>true if a white space separator is placed between genotype at each site.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A sequence alignment as a data matrix. </dd></dl>

</div>
</div>
<a class="anchor" id="a006d5efa47cda07ee4daa9e2d4190f13"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double prob </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the last "prob" read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance. </p>
<p>"prob" is returned by ms when a fixed number of segregating sites is used in conjunction with a theta value. If a "prob" value was present in the last simulation read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance, it will be returned by this method. A value of -1 is returned if no simulation was read, or if the last simulation didn't contain a "prob" value or if the last simulation provoked an exception before reaching the "prob" line. </p>

</div>
</div>
<a class="anchor" id="a466e7f268b9a787a6ba108b29fd76b8d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">double tMRCA </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the last tMRCA read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance. </p>
<p>If a tMRCA value was present in the last simulation read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance, it will be returned by this method. A value of -1. is returned if no simulation was read, or if the last simulation didn't contain a tMRCA value or if the last simulation provoked an exception before reaching the tMRCA line. </p>

</div>
</div>
<a class="anchor" id="aad9932fcd08983abecdec54cb658ba53"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::string trees </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the tree string found in the last simulation read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance. </p>
<p>If one or more trees were present in the last simulation read by any <a class="el" href="classegglib_1_1Ms.html" title="ms-like sequence format parser ">Ms</a> instance, they will be returned as a unique string by this method. An empty string is returned if no simulation was read, or if the last simulation, or if the last simulation didn't contain any tree value or if the last simulation provoked an exception before reaching the tree line.</p>
<p>Note: the trees are returned as a single line. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/<a class="el" href="Ms_8hpp_source.html">Ms.hpp</a></li>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/Ms.cpp</li>
</ul>
</div><!-- contents -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head></head><body><hr size="1">
<address style="text-align: right;"><small>Hosted by&nbsp;</address>
<a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
</body></html>
