<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>EggLib C++ library reference manual</title><link href="doxygen.css" rel="stylesheet" type="text/css">
<head>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname"><a href="../index.html">Egglib</a>&#160;<span id="projectnumber">2.1.11</span></div>
   <div id="projectbrief">C++&#160;library&#160;reference&#160;manual</div>
  </td>
 </tr>
 </tbody>
</table>
</head>
</html>
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>egglib</b></li><li class="navelem"><a class="el" href="classegglib_1_1Mutator.html">Mutator</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classegglib_1_1Mutator-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">Mutator Class Reference<div class="ingroups"><a class="el" href="group__coalesce.html">coalesce</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Implements mutation models.  
 <a href="classegglib_1_1Mutator.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="Mutator_8hpp_source.html">Mutator.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aa2e4d0354e4bae70253f5204f5ba5690"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#aa2e4d0354e4bae70253f5204f5ba5690">Mutator</a> ()</td></tr>
<tr class="memdesc:aa2e4d0354e4bae70253f5204f5ba5690"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initializes with default values.  <a href="#aa2e4d0354e4bae70253f5204f5ba5690">More...</a><br /></td></tr>
<tr class="separator:aa2e4d0354e4bae70253f5204f5ba5690"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a977d52e8cc755fad37c767997d87ec6f"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a977d52e8cc755fad37c767997d87ec6f">~Mutator</a> ()</td></tr>
<tr class="memdesc:a977d52e8cc755fad37c767997d87ec6f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destroys the object.  <a href="#a977d52e8cc755fad37c767997d87ec6f">More...</a><br /></td></tr>
<tr class="separator:a977d52e8cc755fad37c767997d87ec6f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ce92ffc54181a706e1fa8a04065a322"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a5ce92ffc54181a706e1fa8a04065a322">Mutator</a> (const <a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp;)</td></tr>
<tr class="memdesc:a5ce92ffc54181a706e1fa8a04065a322"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy constructor.  <a href="#a5ce92ffc54181a706e1fa8a04065a322">More...</a><br /></td></tr>
<tr class="separator:a5ce92ffc54181a706e1fa8a04065a322"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af83a4d2e005c6545552e3ccdf42a7324"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#af83a4d2e005c6545552e3ccdf42a7324">operator=</a> (const <a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp;)</td></tr>
<tr class="memdesc:af83a4d2e005c6545552e3ccdf42a7324"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assignement operator.  <a href="#af83a4d2e005c6545552e3ccdf42a7324">More...</a><br /></td></tr>
<tr class="separator:af83a4d2e005c6545552e3ccdf42a7324"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad20897c5c8bd47f5d4005989bead0e55"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#ad20897c5c8bd47f5d4005989bead0e55">reset</a> ()</td></tr>
<tr class="memdesc:ad20897c5c8bd47f5d4005989bead0e55"><td class="mdescLeft">&#160;</td><td class="mdescRight">Restores default values of all parameters.  <a href="#ad20897c5c8bd47f5d4005989bead0e55">More...</a><br /></td></tr>
<tr class="separator:ad20897c5c8bd47f5d4005989bead0e55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a6acf4c88994d5966e8c9b5511e2c32"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a0a6acf4c88994d5966e8c9b5511e2c32">fixedNumberOfMutations</a> () const </td></tr>
<tr class="memdesc:a0a6acf4c88994d5966e8c9b5511e2c32"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the fixed number of mutations.  <a href="#a0a6acf4c88994d5966e8c9b5511e2c32">More...</a><br /></td></tr>
<tr class="separator:a0a6acf4c88994d5966e8c9b5511e2c32"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5121410e3f1d81722b1e0d98869316eb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a5121410e3f1d81722b1e0d98869316eb">fixedNumberOfMutations</a> (unsigned int)</td></tr>
<tr class="memdesc:a5121410e3f1d81722b1e0d98869316eb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the fixed number of mutations.  <a href="#a5121410e3f1d81722b1e0d98869316eb">More...</a><br /></td></tr>
<tr class="separator:a5121410e3f1d81722b1e0d98869316eb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab96fb2a8fc075512c839ed3ade681f9"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#aab96fb2a8fc075512c839ed3ade681f9">mutationRate</a> () const </td></tr>
<tr class="memdesc:aab96fb2a8fc075512c839ed3ade681f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the mutation rate.  <a href="#aab96fb2a8fc075512c839ed3ade681f9">More...</a><br /></td></tr>
<tr class="separator:aab96fb2a8fc075512c839ed3ade681f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4aa1f7d0724a640daec9c9d540cb7a1c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a4aa1f7d0724a640daec9c9d540cb7a1c">mutationRate</a> (double)</td></tr>
<tr class="memdesc:a4aa1f7d0724a640daec9c9d540cb7a1c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the mutation rate.  <a href="#a4aa1f7d0724a640daec9c9d540cb7a1c">More...</a><br /></td></tr>
<tr class="separator:a4aa1f7d0724a640daec9c9d540cb7a1c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ec086fc8c14fdc3e55dcfdfe767eae8"><td class="memItemLeft" align="right" valign="top">char&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a6ec086fc8c14fdc3e55dcfdfe767eae8">mutationModel</a> () const </td></tr>
<tr class="memdesc:a6ec086fc8c14fdc3e55dcfdfe767eae8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the mutation model.  <a href="#a6ec086fc8c14fdc3e55dcfdfe767eae8">More...</a><br /></td></tr>
<tr class="separator:a6ec086fc8c14fdc3e55dcfdfe767eae8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59d9b307f7673675e997280271c2d750"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a59d9b307f7673675e997280271c2d750">mutationModel</a> (char)</td></tr>
<tr class="memdesc:a59d9b307f7673675e997280271c2d750"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the mutation model.  <a href="#a59d9b307f7673675e997280271c2d750">More...</a><br /></td></tr>
<tr class="separator:a59d9b307f7673675e997280271c2d750"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb0dbceb536470256c70061e6bca2ff4"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#abb0dbceb536470256c70061e6bca2ff4">numberOfAlleles</a> () const </td></tr>
<tr class="memdesc:abb0dbceb536470256c70061e6bca2ff4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the fixed number of possible alleles.  <a href="#abb0dbceb536470256c70061e6bca2ff4">More...</a><br /></td></tr>
<tr class="separator:abb0dbceb536470256c70061e6bca2ff4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa47ea6c074e5a302519a7d96d85a26b4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#aa47ea6c074e5a302519a7d96d85a26b4">numberOfAlleles</a> (unsigned int)</td></tr>
<tr class="memdesc:aa47ea6c074e5a302519a7d96d85a26b4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the fixed number of possible alleles.  <a href="#aa47ea6c074e5a302519a7d96d85a26b4">More...</a><br /></td></tr>
<tr class="separator:aa47ea6c074e5a302519a7d96d85a26b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace425106b142dec513429b4262801e5e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#ace425106b142dec513429b4262801e5e">transitionWeight</a> (unsigned int i, unsigned int j, double value)</td></tr>
<tr class="memdesc:ace425106b142dec513429b4262801e5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets a transition weight.  <a href="#ace425106b142dec513429b4262801e5e">More...</a><br /></td></tr>
<tr class="separator:ace425106b142dec513429b4262801e5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac13fa2a28e27f875dc973f9d3d0b70a5"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#ac13fa2a28e27f875dc973f9d3d0b70a5">transitionWeight</a> (unsigned int i, unsigned int j)</td></tr>
<tr class="memdesc:ac13fa2a28e27f875dc973f9d3d0b70a5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a transition weight.  <a href="#ac13fa2a28e27f875dc973f9d3d0b70a5">More...</a><br /></td></tr>
<tr class="separator:ac13fa2a28e27f875dc973f9d3d0b70a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc99d704f3439d0dd444664ef581cb8e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#afc99d704f3439d0dd444664ef581cb8e">randomAncestralAllele</a> (bool flag)</td></tr>
<tr class="memdesc:afc99d704f3439d0dd444664ef581cb8e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set to true to draw ancestral alleles randomly.  <a href="#afc99d704f3439d0dd444664ef581cb8e">More...</a><br /></td></tr>
<tr class="separator:afc99d704f3439d0dd444664ef581cb8e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a227b15f1b2c666aba58a6e48ddc00afb"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a227b15f1b2c666aba58a6e48ddc00afb">randomAncestralAllele</a> () const </td></tr>
<tr class="memdesc:a227b15f1b2c666aba58a6e48ddc00afb"><td class="mdescLeft">&#160;</td><td class="mdescRight">true if ancestral alleles must be drawn randomly  <a href="#a227b15f1b2c666aba58a6e48ddc00afb">More...</a><br /></td></tr>
<tr class="separator:a227b15f1b2c666aba58a6e48ddc00afb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c3a305dffa49bc6084a0edb862a73a7"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a7c3a305dffa49bc6084a0edb862a73a7">TPMproba</a> () const </td></tr>
<tr class="memdesc:a7c3a305dffa49bc6084a0edb862a73a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the TPM probability parameter.  <a href="#a7c3a305dffa49bc6084a0edb862a73a7">More...</a><br /></td></tr>
<tr class="separator:a7c3a305dffa49bc6084a0edb862a73a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a083676aa7d61b266504d60f834e9efaa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a083676aa7d61b266504d60f834e9efaa">TPMproba</a> (double value)</td></tr>
<tr class="memdesc:a083676aa7d61b266504d60f834e9efaa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the TPM probability parameter.  <a href="#a083676aa7d61b266504d60f834e9efaa">More...</a><br /></td></tr>
<tr class="separator:a083676aa7d61b266504d60f834e9efaa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56bc70f00b1e2df3d16067a78ae41ed2"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a56bc70f00b1e2df3d16067a78ae41ed2">TPMparam</a> () const </td></tr>
<tr class="memdesc:a56bc70f00b1e2df3d16067a78ae41ed2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the TPM distribution parameter.  <a href="#a56bc70f00b1e2df3d16067a78ae41ed2">More...</a><br /></td></tr>
<tr class="separator:a56bc70f00b1e2df3d16067a78ae41ed2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afa35a6b7406c8f0708a5a163d06b774a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#afa35a6b7406c8f0708a5a163d06b774a">TPMparam</a> (double value)</td></tr>
<tr class="memdesc:afa35a6b7406c8f0708a5a163d06b774a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the TPM distribution parameter.  <a href="#afa35a6b7406c8f0708a5a163d06b774a">More...</a><br /></td></tr>
<tr class="separator:afa35a6b7406c8f0708a5a163d06b774a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af76f81b6f01ba61551471d1e35603e50"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#af76f81b6f01ba61551471d1e35603e50">numberOfSites</a> () const </td></tr>
<tr class="memdesc:af76f81b6f01ba61551471d1e35603e50"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the number of mutable sites.  <a href="#af76f81b6f01ba61551471d1e35603e50">More...</a><br /></td></tr>
<tr class="separator:af76f81b6f01ba61551471d1e35603e50"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa69520df2f40fd7e8d79db718f5e9ae6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#aa69520df2f40fd7e8d79db718f5e9ae6">numberOfSites</a> (unsigned int)</td></tr>
<tr class="memdesc:aa69520df2f40fd7e8d79db718f5e9ae6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the number of mutable sites.  <a href="#aa69520df2f40fd7e8d79db718f5e9ae6">More...</a><br /></td></tr>
<tr class="separator:aa69520df2f40fd7e8d79db718f5e9ae6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9c2c8e6b792fd04679e5ae155d734bb0"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a9c2c8e6b792fd04679e5ae155d734bb0">sitePosition</a> (unsigned int siteIndex) const </td></tr>
<tr class="memdesc:a9c2c8e6b792fd04679e5ae155d734bb0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the position of a given site.  <a href="#a9c2c8e6b792fd04679e5ae155d734bb0">More...</a><br /></td></tr>
<tr class="separator:a9c2c8e6b792fd04679e5ae155d734bb0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc7fea1b11ea75063b77e34350cd123d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#afc7fea1b11ea75063b77e34350cd123d">sitePosition</a> (unsigned int siteIndex, double position)</td></tr>
<tr class="memdesc:afc7fea1b11ea75063b77e34350cd123d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the position of a given site.  <a href="#afc7fea1b11ea75063b77e34350cd123d">More...</a><br /></td></tr>
<tr class="separator:afc7fea1b11ea75063b77e34350cd123d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d36b282ec2fc8b47bdfdc985bf2a8d5"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a2d36b282ec2fc8b47bdfdc985bf2a8d5">siteWeight</a> (unsigned int siteIndex) const </td></tr>
<tr class="memdesc:a2d36b282ec2fc8b47bdfdc985bf2a8d5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the mutation weight of a given site.  <a href="#a2d36b282ec2fc8b47bdfdc985bf2a8d5">More...</a><br /></td></tr>
<tr class="separator:a2d36b282ec2fc8b47bdfdc985bf2a8d5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a15c4585660fe751fdc4310226ab9fa65"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a15c4585660fe751fdc4310226ab9fa65">siteWeight</a> (unsigned int siteIndex, double weight)</td></tr>
<tr class="memdesc:a15c4585660fe751fdc4310226ab9fa65"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the site weight of a given site.  <a href="#a15c4585660fe751fdc4310226ab9fa65">More...</a><br /></td></tr>
<tr class="separator:a15c4585660fe751fdc4310226ab9fa65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a705124e41c038cfe415e966827679db0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a705124e41c038cfe415e966827679db0">mute</a> (<a class="el" href="classegglib_1_1Arg.html">Arg</a> *arg, <a class="el" href="classegglib_1_1Random.html">Random</a> *random)</td></tr>
<tr class="memdesc:a705124e41c038cfe415e966827679db0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs mutation.  <a href="#a705124e41c038cfe415e966827679db0">More...</a><br /></td></tr>
<tr class="separator:a705124e41c038cfe415e966827679db0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3edf62f005feffde936f4ffdfe837b0d"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1Mutator.html#a3edf62f005feffde936f4ffdfe837b0d">numberOfMutations</a> () const </td></tr>
<tr class="memdesc:a3edf62f005feffde936f4ffdfe837b0d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the last number of mutations.  <a href="#a3edf62f005feffde936f4ffdfe837b0d">More...</a><br /></td></tr>
<tr class="separator:a3edf62f005feffde936f4ffdfe837b0d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Implements mutation models. </p>
<p>Works with a previously built Ancestral Reconbination Graph. The user must sets options using the setter-based interface. After that he or she can call the method <a class="el" href="classegglib_1_1Mutator.html#a705124e41c038cfe415e966827679db0" title="Performs mutation. ">mute()</a> that will generates a <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> object.</p>
<p>Genotype data are represented by integer numbers. Regardless of the mutation model, the ancestral state is always 0. The user can set the rate of mutation (or, alternatively, fix the number of mutations that occurred - which is the number of segregating sites only with an infinite site model).</p>
<p>Other options fall into two separate groups: the positions of the mutated sites and the process of mutation (how new alleles are generated).</p>
<p>Concerning allele generation, several mutation models are available (coded by single letters):</p><ul>
<li>F: fixed number of alleles. Among other markers, this model is appropriate for simulating nucleotides. The user is able to choose the number of alleles (where 2 is the standard for an infinite site model and 4 for a finite site model). <a class="el" href="classegglib_1_1Mutator.html" title="Implements mutation models. ">Mutator</a> allows assigning independent weights between all different transition types and can draw randomly the ancestral states, providing a way to emulate evolution of nucleotides with multiple mutations at the same site and reversion.</li>
<li>I: infinite number of alleles. At a given site, each mutation raises a new allele. The value of the alleles is therefore irrelevant (it only denotes its order of appearance). This model does not permit homoplasy.</li>
<li>S: stepwise mutation model. In this model the value of the alleles are interpreted as a size (typically for simulating a microsatellite marker). Each mutation either increases or decreases the allele size by an increment of one.</li>
<li>T: two-phase mutation model. This model is a generalization of the stepwise mutation model (S). For a mutation, the increment (either increase or decrease) is 1 with the probability given by the parameter (1-TPMproba). With probability TPMproba, the increment is drawn from a geometric distribution of parameter given by the other parameter (TPMparam).</li>
</ul>
<p>By default, the program will assume an infinite site model (ISM). Each mutation will occur to a new position drawn from the interval [0,1]. It is possible to set any mutation model with an ISM (including microsatellite-like models I, S and T). Alternatively, the user can specify a finite number of sites available for mutation. For a microsatellite marker, the user will want to specify a single site. It is possible to set a finite number of sites for all mutation models. In all cases, the mutations will be forced to target these sites. It is possible to apply weights independently to all sites. The higher the weight value (comparatively to the other sites), the higher the probability that this site mutes. The weights needs not to be relative. In addition, the user can set the positions of the different sites. Nothings forces him or her to place them in order. Note that this does not affect the mutation process, but on the amount of recombination that will be allowed between sites. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="aa2e4d0354e4bae70253f5204f5ba5690"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Mutator.html">Mutator</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initializes with default values. </p>
<p>List of default values:</p><ul>
<li>theta = 0</li>
<li>fixedNumberOfMutations = 0</li>
<li>model = F (fixed number of alleles)</li>
<li>fixed number of alleles = 2</li>
<li>infinite site model</li>
<li>TPM parameters are both preset to 0.5 </li>
</ul>

</div>
</div>
<a class="anchor" id="a977d52e8cc755fad37c767997d87ec6f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">~<a class="el" href="classegglib_1_1Mutator.html">Mutator</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Destroys the object. </p>

</div>
</div>
<a class="anchor" id="a5ce92ffc54181a706e1fa8a04065a322"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Mutator.html">Mutator</a> </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp;&#160;</td>
          <td class="paramname"><em>src</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Copy constructor. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a0a6acf4c88994d5966e8c9b5511e2c32"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int fixedNumberOfMutations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the fixed number of mutations. </p>

</div>
</div>
<a class="anchor" id="a5121410e3f1d81722b1e0d98869316eb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void fixedNumberOfMutations </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the fixed number of mutations. </p>
<p>The value can be 0. It is not allowed to set both the fixed number of mutations and the mutation rate at non-zero value </p>

</div>
</div>
<a class="anchor" id="a6ec086fc8c14fdc3e55dcfdfe767eae8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char mutationModel </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the mutation model. </p>
<p>See the class documentation for the signification of the different one-letter codes. </p>

</div>
</div>
<a class="anchor" id="a59d9b307f7673675e997280271c2d750"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void mutationModel </td>
          <td>(</td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the mutation model. </p>
<p>The passed character must be one of F, I, S and T. See the class documentation for their signification. </p>

</div>
</div>
<a class="anchor" id="aab96fb2a8fc075512c839ed3ade681f9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double mutationRate </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the mutation rate. </p>

</div>
</div>
<a class="anchor" id="a4aa1f7d0724a640daec9c9d540cb7a1c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void mutationRate </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the mutation rate. </p>
<p>The value cannot be negative. The value can be 0. It is not allowed to set both the fixed number of mutations and the mutation rate at non-zero value </p>

</div>
</div>
<a class="anchor" id="a705124e41c038cfe415e966827679db0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1DataMatrix.html">DataMatrix</a> mute </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1Arg.html">Arg</a> *&#160;</td>
          <td class="paramname"><em>arg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classegglib_1_1Random.html">Random</a> *&#160;</td>
          <td class="paramname"><em>random</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs mutation. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">arg</td><td>Ancestral recombination graph instance. If the ARG is partially built or not a all, or improperly so, the behaviour of this method is not defined.</td></tr>
    <tr><td class="paramname">random</td><td>The address of a <a class="el" href="classegglib_1_1Random.html" title="Pseudo-random number generator. ">Random</a> instance to be used for generating random numbers.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="classegglib_1_1DataMatrix.html" title="Data table. ">DataMatrix</a> instance containing simulated data. </dd></dl>

</div>
</div>
<a class="anchor" id="abb0dbceb536470256c70061e6bca2ff4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int numberOfAlleles </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the fixed number of possible alleles. </p>

</div>
</div>
<a class="anchor" id="aa47ea6c074e5a302519a7d96d85a26b4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void numberOfAlleles </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the fixed number of possible alleles. </p>
<p>The value must be larger than 1. This parameter is meaningful only for the fixed number allele model of mutation, and ignored otherwise. </p>

</div>
</div>
<a class="anchor" id="a3edf62f005feffde936f4ffdfe837b0d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int numberOfMutations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the last number of mutations. </p>
<p>Returns the number of mutations of the last call of <a class="el" href="classegglib_1_1Mutator.html#a705124e41c038cfe415e966827679db0" title="Performs mutation. ">mute( )</a>. By default, this method returns 0. </p>

</div>
</div>
<a class="anchor" id="af76f81b6f01ba61551471d1e35603e50"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int numberOfSites </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the number of mutable sites. </p>
<p>A value a zero must be interpreted as the infinite site model. Note that after all calls all data from the tables sitePositions and siteWeights will be reset. </p>

</div>
</div>
<a class="anchor" id="aa69520df2f40fd7e8d79db718f5e9ae6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void numberOfSites </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the number of mutable sites. </p>
<p>The value of zero is accepted and imposed the infinite site model. </p>

</div>
</div>
<a class="anchor" id="af83a4d2e005c6545552e3ccdf42a7324"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp; operator= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1Mutator.html">Mutator</a> &amp;&#160;</td>
          <td class="paramname"><em>src</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Assignement operator. </p>

</div>
</div>
<a class="anchor" id="afc99d704f3439d0dd444664ef581cb8e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void randomAncestralAllele </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>flag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set to true to draw ancestral alleles randomly. </p>
<p>By default, the ancestral allele is always 0. If this variable is set to true, the ancestral allele will be randomly drawn from the defined number of alleles. This option is always ignored unless in combination with the Fixed Allele Number model. </p>

</div>
</div>
<a class="anchor" id="a227b15f1b2c666aba58a6e48ddc00afb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool randomAncestralAllele </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>true if ancestral alleles must be drawn randomly </p>

</div>
</div>
<a class="anchor" id="ad20897c5c8bd47f5d4005989bead0e55"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void reset </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Restores default values of all parameters. </p>

</div>
</div>
<a class="anchor" id="a9c2c8e6b792fd04679e5ae155d734bb0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double sitePosition </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>siteIndex</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the position of a given site. </p>

</div>
</div>
<a class="anchor" id="afc7fea1b11ea75063b77e34350cd123d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void sitePosition </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>siteIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>position</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set the position of a given site. </p>
<p>The position must be &gt;=0 and &lt;=1 </p>

</div>
</div>
<a class="anchor" id="a2d36b282ec2fc8b47bdfdc985bf2a8d5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double siteWeight </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>siteIndex</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the mutation weight of a given site. </p>

</div>
</div>
<a class="anchor" id="a15c4585660fe751fdc4310226ab9fa65"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void siteWeight </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>siteIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>weight</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set the site weight of a given site. </p>
<p>The weight must be strictly positive. </p>

</div>
</div>
<a class="anchor" id="a56bc70f00b1e2df3d16067a78ae41ed2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double TPMparam </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the TPM distribution parameter. </p>

</div>
</div>
<a class="anchor" id="afa35a6b7406c8f0708a5a163d06b774a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void TPMparam </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the TPM distribution parameter. </p>
<p>This parameter is considered only if the mutation model is T (two-phase mutation model). It gives the parameter of the geometric distribution which is used to generate the mutation step (if it is not one).</p>
<p>The value must be &gt;=0. and &lt;=1. </p>

</div>
</div>
<a class="anchor" id="a7c3a305dffa49bc6084a0edb862a73a7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double TPMproba </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the TPM probability parameter. </p>

</div>
</div>
<a class="anchor" id="a083676aa7d61b266504d60f834e9efaa"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void TPMproba </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets the TPM probability parameter. </p>
<p>This parameter is considered only if the mutation model is T (two-phase mutation model). It gives the probability that a mutation step is not fixed to be 1. If TPMproba is 0, the mutation model is SMM.</p>
<p>The value must be &gt;=0. and &lt;=1. </p>

</div>
</div>
<a class="anchor" id="ace425106b142dec513429b4262801e5e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void transitionWeight </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>j</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>value</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sets a transition weight. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>row (previous allele index). </td></tr>
    <tr><td class="paramname">j</td><td>column (next allele index). </td></tr>
    <tr><td class="paramname">value</td><td>weight to apply.</td></tr>
  </table>
  </dd>
</dl>
<p>Indices i and j must be different. Weights can be any strictly positive value. </p>

</div>
</div>
<a class="anchor" id="ac13fa2a28e27f875dc973f9d3d0b70a5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double transitionWeight </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets a transition weight. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>row (previous allele index). </td></tr>
    <tr><td class="paramname">j</td><td>column (next allele index).</td></tr>
  </table>
  </dd>
</dl>
<p>Indices i and j must be different. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/<a class="el" href="Mutator_8hpp_source.html">Mutator.hpp</a></li>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/Mutator.cpp</li>
</ul>
</div><!-- contents -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head></head><body><hr size="1">
<address style="text-align: right;"><small>Hosted by&nbsp;</address>
<a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
</body></html>
