<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
<title>EggLib C++ library reference manual</title><link href="doxygen.css" rel="stylesheet" type="text/css">
<head>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css"/>
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname"><a href="../index.html">Egglib</a>&#160;<span id="projectnumber">2.1.11</span></div>
   <div id="projectbrief">C++&#160;library&#160;reference&#160;manual</div>
  </td>
 </tr>
 </tbody>
</table>
</head>
</html>
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>egglib</b></li><li class="navelem"><a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classegglib_1_1NucleotideDiversity-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">NucleotideDiversity Class Reference<div class="ingroups"><a class="el" href="group__polymorphism.html">polymorphism</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Performs analyzes of population genetics.  
 <a href="classegglib_1_1NucleotideDiversity.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="NucleotideDiversity_8hpp_source.html">NucleotideDiversity.hpp</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for NucleotideDiversity:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classegglib_1_1NucleotideDiversity.png" usemap="#NucleotideDiversity_map" alt=""/>
  <map id="NucleotideDiversity_map" name="NucleotideDiversity_map">
<area href="classegglib_1_1BaseDiversity.html" title="Base class of diversity classes. " alt="BaseDiversity" shape="rect" coords="0,0,122,24"/>
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a4b81fe7e6f785b5413d9bfd192fa39e3"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a4b81fe7e6f785b5413d9bfd192fa39e3">NucleotideDiversity</a> ()</td></tr>
<tr class="memdesc:a4b81fe7e6f785b5413d9bfd192fa39e3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Builds an object.  <a href="#a4b81fe7e6f785b5413d9bfd192fa39e3">More...</a><br /></td></tr>
<tr class="separator:a4b81fe7e6f785b5413d9bfd192fa39e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aadecb2de0968a245ae1b36543bdec89b"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#aadecb2de0968a245ae1b36543bdec89b">~NucleotideDiversity</a> ()</td></tr>
<tr class="memdesc:aadecb2de0968a245ae1b36543bdec89b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destroys an object.  <a href="#aadecb2de0968a245ae1b36543bdec89b">More...</a><br /></td></tr>
<tr class="separator:aadecb2de0968a245ae1b36543bdec89b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f7781582853b62c254596290a0c0af1"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a0f7781582853b62c254596290a0c0af1">load</a> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, bool allowMultipleMutations=false, double minimumExploitableData=1., unsigned int ignoreFrequency=0, std::string characterMapping=<a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a>, bool useZeroAsAncestral=false)</td></tr>
<tr class="memdesc:a0f7781582853b62c254596290a0c0af1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identifies polymorphic sites and computes basis statistics.  <a href="#a0f7781582853b62c254596290a0c0af1">More...</a><br /></td></tr>
<tr class="separator:a0f7781582853b62c254596290a0c0af1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a807ed7df5b0b1f808ebe6c1c4dfe3232"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a807ed7df5b0b1f808ebe6c1c4dfe3232"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a807ed7df5b0b1f808ebe6c1c4dfe3232">S</a> () const </td></tr>
<tr class="memdesc:a807ed7df5b0b1f808ebe6c1c4dfe3232"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of polymorphic sites. <br /></td></tr>
<tr class="separator:a807ed7df5b0b1f808ebe6c1c4dfe3232"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac7cb98ed1f63445af2504630c7061f13"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac7cb98ed1f63445af2504630c7061f13"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ac7cb98ed1f63445af2504630c7061f13">So</a> () const </td></tr>
<tr class="memdesc:ac7cb98ed1f63445af2504630c7061f13"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of polymorphic orientable sites. <br /></td></tr>
<tr class="separator:ac7cb98ed1f63445af2504630c7061f13"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a490e00b1bd94d024745e6aa2ea1ad072"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a490e00b1bd94d024745e6aa2ea1ad072"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a490e00b1bd94d024745e6aa2ea1ad072">eta</a> () const </td></tr>
<tr class="memdesc:a490e00b1bd94d024745e6aa2ea1ad072"><td class="mdescLeft">&#160;</td><td class="mdescRight">Minimum number of mutations. <br /></td></tr>
<tr class="separator:a490e00b1bd94d024745e6aa2ea1ad072"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0e0b380f228e728114d8b3614ead7bcd"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0e0b380f228e728114d8b3614ead7bcd"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a0e0b380f228e728114d8b3614ead7bcd">nseff</a> () const </td></tr>
<tr class="memdesc:a0e0b380f228e728114d8b3614ead7bcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Average of per-site number of sequences effectively used. <br /></td></tr>
<tr class="separator:a0e0b380f228e728114d8b3614ead7bcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1892aa2408d56bc0956b48a1f60c448a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1892aa2408d56bc0956b48a1f60c448a"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a1892aa2408d56bc0956b48a1f60c448a">lseff</a> () const </td></tr>
<tr class="memdesc:a1892aa2408d56bc0956b48a1f60c448a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites effectively used. <br /></td></tr>
<tr class="separator:a1892aa2408d56bc0956b48a1f60c448a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a60dff9390a139542033e5f8850a256a4"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a60dff9390a139542033e5f8850a256a4"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a60dff9390a139542033e5f8850a256a4">nseffo</a> () const </td></tr>
<tr class="memdesc:a60dff9390a139542033e5f8850a256a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Average of number of sequences effectively used at orientable sites. <br /></td></tr>
<tr class="separator:a60dff9390a139542033e5f8850a256a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7a2c1cd74867167680a5e8c7aaeb8dc9"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7a2c1cd74867167680a5e8c7aaeb8dc9"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a7a2c1cd74867167680a5e8c7aaeb8dc9">lseffo</a> () const </td></tr>
<tr class="memdesc:a7a2c1cd74867167680a5e8c7aaeb8dc9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of orientable sites. <br /></td></tr>
<tr class="separator:a7a2c1cd74867167680a5e8c7aaeb8dc9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a64e2a336c5528a16a2003212b7d737ad"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a64e2a336c5528a16a2003212b7d737ad"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a64e2a336c5528a16a2003212b7d737ad">npop</a> () const </td></tr>
<tr class="memdesc:a64e2a336c5528a16a2003212b7d737ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of detected populations. <br /></td></tr>
<tr class="separator:a64e2a336c5528a16a2003212b7d737ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4d828106c69b4365b51f4ebb50fcaf2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac4d828106c69b4365b51f4ebb50fcaf2"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ac4d828106c69b4365b51f4ebb50fcaf2">popLabel</a> (unsigned int popIndex) const </td></tr>
<tr class="memdesc:ac4d828106c69b4365b51f4ebb50fcaf2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Label of the population with given index (unsecure) <br /></td></tr>
<tr class="separator:ac4d828106c69b4365b51f4ebb50fcaf2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb3a149b8b64b8dea4acdf0e760f8936"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adb3a149b8b64b8dea4acdf0e760f8936"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#adb3a149b8b64b8dea4acdf0e760f8936">Pi</a> ()</td></tr>
<tr class="memdesc:adb3a149b8b64b8dea4acdf0e760f8936"><td class="mdescLeft">&#160;</td><td class="mdescRight">Nucleotide diversity. <br /></td></tr>
<tr class="separator:adb3a149b8b64b8dea4acdf0e760f8936"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a81dd9a908d49ab69c9acc4ebfd02b910"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a81dd9a908d49ab69c9acc4ebfd02b910"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a81dd9a908d49ab69c9acc4ebfd02b910">thetaW</a> ()</td></tr>
<tr class="memdesc:a81dd9a908d49ab69c9acc4ebfd02b910"><td class="mdescLeft">&#160;</td><td class="mdescRight">Watterson estimator of theta. <br /></td></tr>
<tr class="separator:a81dd9a908d49ab69c9acc4ebfd02b910"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3375e5e80f1fc4d46c3c91ab29044074"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3375e5e80f1fc4d46c3c91ab29044074"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a3375e5e80f1fc4d46c3c91ab29044074">average_Pi</a> ()</td></tr>
<tr class="memdesc:a3375e5e80f1fc4d46c3c91ab29044074"><td class="mdescLeft">&#160;</td><td class="mdescRight">Average of Pi over populations. <br /></td></tr>
<tr class="separator:a3375e5e80f1fc4d46c3c91ab29044074"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a43ce6dda0b6fa4a43b103966f9c0362f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a43ce6dda0b6fa4a43b103966f9c0362f"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a43ce6dda0b6fa4a43b103966f9c0362f">pop_Pi</a> (unsigned int popIndex)</td></tr>
<tr class="memdesc:a43ce6dda0b6fa4a43b103966f9c0362f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Pi of a given population (unsecure) <br /></td></tr>
<tr class="separator:a43ce6dda0b6fa4a43b103966f9c0362f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a996cc5ffcedfeb9d102bea52a646bb33"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a996cc5ffcedfeb9d102bea52a646bb33"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a996cc5ffcedfeb9d102bea52a646bb33">D</a> ()</td></tr>
<tr class="memdesc:a996cc5ffcedfeb9d102bea52a646bb33"><td class="mdescLeft">&#160;</td><td class="mdescRight">Tajima's D. <br /></td></tr>
<tr class="separator:a996cc5ffcedfeb9d102bea52a646bb33"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1884acece8703c603eb1d17bffd551f8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1884acece8703c603eb1d17bffd551f8"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a1884acece8703c603eb1d17bffd551f8">thetaH</a> ()</td></tr>
<tr class="memdesc:a1884acece8703c603eb1d17bffd551f8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fay and Wu estimator of theta. <br /></td></tr>
<tr class="separator:a1884acece8703c603eb1d17bffd551f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ac299db977317d29f8b60a36f7c2def"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2ac299db977317d29f8b60a36f7c2def"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a2ac299db977317d29f8b60a36f7c2def">thetaL</a> ()</td></tr>
<tr class="memdesc:a2ac299db977317d29f8b60a36f7c2def"><td class="mdescLeft">&#160;</td><td class="mdescRight">Zeng et al. estimator of theta. <br /></td></tr>
<tr class="separator:a2ac299db977317d29f8b60a36f7c2def"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae2998d33cf5bea75c723690a0605770b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae2998d33cf5bea75c723690a0605770b"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ae2998d33cf5bea75c723690a0605770b">H</a> ()</td></tr>
<tr class="memdesc:ae2998d33cf5bea75c723690a0605770b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fay and Wu's H. <br /></td></tr>
<tr class="separator:ae2998d33cf5bea75c723690a0605770b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2bd27ccb8f185c0e84af4a0762dbbc7e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2bd27ccb8f185c0e84af4a0762dbbc7e"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a2bd27ccb8f185c0e84af4a0762dbbc7e">Z</a> ()</td></tr>
<tr class="memdesc:a2bd27ccb8f185c0e84af4a0762dbbc7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Standardized H. <br /></td></tr>
<tr class="separator:a2bd27ccb8f185c0e84af4a0762dbbc7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a31a72294fbea2f70ab4c6eff07f7ce37"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a31a72294fbea2f70ab4c6eff07f7ce37"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a31a72294fbea2f70ab4c6eff07f7ce37">E</a> ()</td></tr>
<tr class="memdesc:a31a72294fbea2f70ab4c6eff07f7ce37"><td class="mdescLeft">&#160;</td><td class="mdescRight">Zeng et al.'s E. <br /></td></tr>
<tr class="separator:a31a72294fbea2f70ab4c6eff07f7ce37"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8541259a9497fa07594b91a3ac8963c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8541259a9497fa07594b91a3ac8963c"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ac8541259a9497fa07594b91a3ac8963c">FixedDifferences</a> ()</td></tr>
<tr class="memdesc:ac8541259a9497fa07594b91a3ac8963c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites with at least one fixed difference. <br /></td></tr>
<tr class="separator:ac8541259a9497fa07594b91a3ac8963c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af747f99e9aa9431d10dd5d82f48d4f54"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af747f99e9aa9431d10dd5d82f48d4f54"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#af747f99e9aa9431d10dd5d82f48d4f54">CommonAlleles</a> ()</td></tr>
<tr class="memdesc:af747f99e9aa9431d10dd5d82f48d4f54"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites with at least one allele shared among at least two populations. <br /></td></tr>
<tr class="separator:af747f99e9aa9431d10dd5d82f48d4f54"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96ad6359db6ad12c0efc87aa08683a5e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a96ad6359db6ad12c0efc87aa08683a5e"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a96ad6359db6ad12c0efc87aa08683a5e">SharedAlleles</a> ()</td></tr>
<tr class="memdesc:a96ad6359db6ad12c0efc87aa08683a5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites with at least one non-fixed allele shared among at least two populations. <br /></td></tr>
<tr class="separator:a96ad6359db6ad12c0efc87aa08683a5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adeeec1f26d1d90d7c5045e957610f63b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adeeec1f26d1d90d7c5045e957610f63b"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#adeeec1f26d1d90d7c5045e957610f63b">SpecificAlleles</a> ()</td></tr>
<tr class="memdesc:adeeec1f26d1d90d7c5045e957610f63b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites with at least one allele specific to one population. <br /></td></tr>
<tr class="separator:adeeec1f26d1d90d7c5045e957610f63b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b19a81ba68d5292da23fc4e488f03b0"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3b19a81ba68d5292da23fc4e488f03b0"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a3b19a81ba68d5292da23fc4e488f03b0">SpecificDerivedAlleles</a> ()</td></tr>
<tr class="memdesc:a3b19a81ba68d5292da23fc4e488f03b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of sites with at least one derived allele specific to one population. <br /></td></tr>
<tr class="separator:a3b19a81ba68d5292da23fc4e488f03b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7670ad548cfc81eda7f733b25f8c0c8a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7670ad548cfc81eda7f733b25f8c0c8a"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a7670ad548cfc81eda7f733b25f8c0c8a">Polymorphisms</a> (unsigned int pop)</td></tr>
<tr class="memdesc:a7670ad548cfc81eda7f733b25f8c0c8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of polymorphisms in a given population (unsecure) <br /></td></tr>
<tr class="separator:a7670ad548cfc81eda7f733b25f8c0c8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1922e1c3e6d49ad5fe80a4df03de6b50"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1922e1c3e6d49ad5fe80a4df03de6b50"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a1922e1c3e6d49ad5fe80a4df03de6b50">SpecificAlleles</a> (unsigned int pop)</td></tr>
<tr class="memdesc:a1922e1c3e6d49ad5fe80a4df03de6b50"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of specific alleles for a given population (unsecure) <br /></td></tr>
<tr class="separator:a1922e1c3e6d49ad5fe80a4df03de6b50"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0fca91c5ac7c38740ce5ae4f0e86396"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad0fca91c5ac7c38740ce5ae4f0e86396"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ad0fca91c5ac7c38740ce5ae4f0e86396">SpecificDerivedAlleles</a> (unsigned int pop)</td></tr>
<tr class="memdesc:ad0fca91c5ac7c38740ce5ae4f0e86396"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of specific derived allele for a given population (unsecure) <br /></td></tr>
<tr class="separator:ad0fca91c5ac7c38740ce5ae4f0e86396"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0a3435a84751cbbb70cf0c8864d6736"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab0a3435a84751cbbb70cf0c8864d6736"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#ab0a3435a84751cbbb70cf0c8864d6736">FixedDifferences</a> (unsigned int pop1, unsigned int pop2)</td></tr>
<tr class="memdesc:ab0a3435a84751cbbb70cf0c8864d6736"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of fixed differences between a given pair of populations (unsecure; pop2 must be larger than pop1) <br /></td></tr>
<tr class="separator:ab0a3435a84751cbbb70cf0c8864d6736"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa4eed62f4c416672ee8955eaf865d34e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa4eed62f4c416672ee8955eaf865d34e"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#aa4eed62f4c416672ee8955eaf865d34e">CommonAlleles</a> (unsigned int pop1, unsigned int pop2)</td></tr>
<tr class="memdesc:aa4eed62f4c416672ee8955eaf865d34e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of common alleles between a given pair of populations (unsecure; pop2 must be larger than pop1) <br /></td></tr>
<tr class="separator:aa4eed62f4c416672ee8955eaf865d34e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a693b04073db17098bc9cc26d1d6c4882"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a693b04073db17098bc9cc26d1d6c4882"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a693b04073db17098bc9cc26d1d6c4882">SharedAlleles</a> (unsigned int pop1, unsigned int pop2)</td></tr>
<tr class="memdesc:a693b04073db17098bc9cc26d1d6c4882"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of shared non-fixed alleles between a given pair of populations (unsecure; pop2 must be larger than pop1) <br /></td></tr>
<tr class="separator:a693b04073db17098bc9cc26d1d6c4882"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a011ed8f7cd114c5665a3444ebd68a24f"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a011ed8f7cd114c5665a3444ebd68a24f">triConfiguration</a> (unsigned int index)</td></tr>
<tr class="memdesc:a011ed8f7cd114c5665a3444ebd68a24f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number falling into one of the possible site configurations.  <a href="#a011ed8f7cd114c5665a3444ebd68a24f">More...</a><br /></td></tr>
<tr class="separator:a011ed8f7cd114c5665a3444ebd68a24f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a934780da36dd6c4eb6f7556b18ed2c12"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a934780da36dd6c4eb6f7556b18ed2c12"></a>
std::vector&lt; unsigned int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a934780da36dd6c4eb6f7556b18ed2c12">polymorphic_positions</a> () const </td></tr>
<tr class="memdesc:a934780da36dd6c4eb6f7556b18ed2c12"><td class="mdescLeft">&#160;</td><td class="mdescRight">Builds and returns the vector of positions of all polymorphic sites. <br /></td></tr>
<tr class="separator:a934780da36dd6c4eb6f7556b18ed2c12"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adc71bbf66c1d95cc029c03e811457825"><td class="memItemLeft" align="right" valign="top">std::vector&lt; unsigned int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#adc71bbf66c1d95cc029c03e811457825">singleton_positions</a> () const </td></tr>
<tr class="memdesc:adc71bbf66c1d95cc029c03e811457825"><td class="mdescLeft">&#160;</td><td class="mdescRight">Builds and returns the vector of positions of all singleton sites.  <a href="#adc71bbf66c1d95cc029c03e811457825">More...</a><br /></td></tr>
<tr class="separator:adc71bbf66c1d95cc029c03e811457825"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a728098ee7906f24ec76a07f88d848a65">BaseDiversity</a> ()</td></tr>
<tr class="memdesc:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor.  <a href="#a728098ee7906f24ec76a07f88d848a65">More...</a><br /></td></tr>
<tr class="separator:a728098ee7906f24ec76a07f88d848a65 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a2155869a8e19b5d53b6d3f4eabb9de42">~BaseDiversity</a> ()</td></tr>
<tr class="memdesc:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor.  <a href="#a2155869a8e19b5d53b6d3f4eabb9de42">More...</a><br /></td></tr>
<tr class="separator:a2155869a8e19b5d53b6d3f4eabb9de42 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#af9121ee3a3a8722b2d36961a802e8189">reserve</a> (unsigned int numberOfSites)</td></tr>
<tr class="memdesc:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reserve sufficient memory for a given number of polymorphic sites.  <a href="#af9121ee3a3a8722b2d36961a802e8189">More...</a><br /></td></tr>
<tr class="separator:af9121ee3a3a8722b2d36961a802e8189 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a77fd109923331c4c39106ca1b99d5e57"></a>
const <a class="el" href="classegglib_1_1SitePolymorphism.html">SitePolymorphism</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a77fd109923331c4c39106ca1b99d5e57">get_site</a> (unsigned int index) const </td></tr>
<tr class="memdesc:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a site. <br /></td></tr>
<tr class="separator:a77fd109923331c4c39106ca1b99d5e57 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2d827860e77ed18cb0356640561abf2f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a2d827860e77ed18cb0356640561abf2f">get_position</a> (unsigned int index) const </td></tr>
<tr class="memdesc:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a site position. <br /></td></tr>
<tr class="separator:a2d827860e77ed18cb0356640561abf2f inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad20897c5c8bd47f5d4005989bead0e55"></a>
virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#ad20897c5c8bd47f5d4005989bead0e55">reset</a> ()</td></tr>
<tr class="memdesc:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears and re-initializes object. <br /></td></tr>
<tr class="separator:ad20897c5c8bd47f5d4005989bead0e55 inherit pub_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:abd422da68dace0766ec6df73032b0974"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#abd422da68dace0766ec6df73032b0974">NucleotideDiversity</a> (const <a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> &amp;source)</td></tr>
<tr class="memdesc:abd422da68dace0766ec6df73032b0974"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be copied.  <a href="#abd422da68dace0766ec6df73032b0974">More...</a><br /></td></tr>
<tr class="separator:abd422da68dace0766ec6df73032b0974"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a426e6e4a99cfc9c89887f24a28af9d3e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1NucleotideDiversity.html#a426e6e4a99cfc9c89887f24a28af9d3e">operator=</a> (const <a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> &amp;source)</td></tr>
<tr class="memdesc:a426e6e4a99cfc9c89887f24a28af9d3e"><td class="mdescLeft">&#160;</td><td class="mdescRight">This class cannot be copied.  <a href="#a426e6e4a99cfc9c89887f24a28af9d3e">More...</a><br /></td></tr>
<tr class="separator:a426e6e4a99cfc9c89887f24a28af9d3e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a02fd73d861ef2e4aabb38c0c9ff82947"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a02fd73d861ef2e4aabb38c0c9ff82947"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>init</b> ()</td></tr>
<tr class="separator:a02fd73d861ef2e4aabb38c0c9ff82947"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8bb3912a3ce86b15842e79d0b421204"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8bb3912a3ce86b15842e79d0b421204"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>clear</b> ()</td></tr>
<tr class="separator:ac8bb3912a3ce86b15842e79d0b421204"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a032a0d82d468d571a2cc494c91ebe3b3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a032a0d82d468d571a2cc494c91ebe3b3"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>diversity</b> ()</td></tr>
<tr class="separator:a032a0d82d468d571a2cc494c91ebe3b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a57b361c0f7cc46ba58acc842b8656b5b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a57b361c0f7cc46ba58acc842b8656b5b"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>outgroupDiversity</b> ()</td></tr>
<tr class="separator:a57b361c0f7cc46ba58acc842b8656b5b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9bd19c8b33de82df1ebdbd4e53676cd0"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9bd19c8b33de82df1ebdbd4e53676cd0"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>differentiation</b> ()</td></tr>
<tr class="separator:a9bd19c8b33de82df1ebdbd4e53676cd0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e8ffce4f659ccfa0e81c280fdd08e30"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4e8ffce4f659ccfa0e81c280fdd08e30"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>triConfigurations</b> ()</td></tr>
<tr class="separator:a4e8ffce4f659ccfa0e81c280fdd08e30"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_methods_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pro_methods_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Protected Member Functions inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a9a7c8b0679c8aa67a65808ea073688a9 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9a7c8b0679c8aa67a65808ea073688a9"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>importSites</b> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, bool allowMultipleMutations, double minimumExploitableData, unsigned int ignoreFrequency, std::string characterMapping, bool useZeroAsAncestral, bool ignoreOutgroup)</td></tr>
<tr class="separator:a9a7c8b0679c8aa67a65808ea073688a9 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af43dec4a12ccce2c941edebcccd6c36a inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af43dec4a12ccce2c941edebcccd6c36a"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>analyzeSite</b> (<a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;data, unsigned int index, double maxMissingData, bool ignoreOutgroup)</td></tr>
<tr class="separator:af43dec4a12ccce2c941edebcccd6c36a inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b4928f332a83517fe59a8624136a249 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8b4928f332a83517fe59a8624136a249"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>getPopIndex</b> (unsigned int label) const </td></tr>
<tr class="separator:a8b4928f332a83517fe59a8624136a249 inherit pro_methods_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:aa87d2f0afec7096dde5c93427c57b294"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa87d2f0afec7096dde5c93427c57b294"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>b_analysisSites</b></td></tr>
<tr class="separator:aa87d2f0afec7096dde5c93427c57b294"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aefa8350c6ee040d887c4a58ae142f274"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aefa8350c6ee040d887c4a58ae142f274"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>b_diversity</b></td></tr>
<tr class="separator:aefa8350c6ee040d887c4a58ae142f274"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac31b0878c26f3d0af2c7ff8d54078122"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac31b0878c26f3d0af2c7ff8d54078122"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_Pi</b></td></tr>
<tr class="separator:ac31b0878c26f3d0af2c7ff8d54078122"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aecab159e253161e8417ec7c4027c1018"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aecab159e253161e8417ec7c4027c1018"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_thetaW</b></td></tr>
<tr class="separator:aecab159e253161e8417ec7c4027c1018"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0639723961abec0f191f53937e09f323"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0639723961abec0f191f53937e09f323"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_average_Pi</b></td></tr>
<tr class="separator:a0639723961abec0f191f53937e09f323"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae9e2c98eded080dc2db58a7cda065ac3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae9e2c98eded080dc2db58a7cda065ac3"></a>
double *&#160;</td><td class="memItemRight" valign="bottom"><b>v_pop_Pi</b></td></tr>
<tr class="separator:ae9e2c98eded080dc2db58a7cda065ac3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a98df34c16a30702508d3c382fc852087"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a98df34c16a30702508d3c382fc852087"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_D</b></td></tr>
<tr class="separator:a98df34c16a30702508d3c382fc852087"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af04c39b164cafde8652518d272565629"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af04c39b164cafde8652518d272565629"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>b_outgroupDiversity</b></td></tr>
<tr class="separator:af04c39b164cafde8652518d272565629"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0594dc03901a847aea1e48390bcd9f10"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0594dc03901a847aea1e48390bcd9f10"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_thetaH</b></td></tr>
<tr class="separator:a0594dc03901a847aea1e48390bcd9f10"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6a9704381e025e0bddcbf8395a24e8b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac6a9704381e025e0bddcbf8395a24e8b"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_thetaL</b></td></tr>
<tr class="separator:ac6a9704381e025e0bddcbf8395a24e8b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5318ef39d131e199e9365c57aaac5d05"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5318ef39d131e199e9365c57aaac5d05"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_H</b></td></tr>
<tr class="separator:a5318ef39d131e199e9365c57aaac5d05"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a62ac5c497b8aeae08e9f3a9789932e4e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a62ac5c497b8aeae08e9f3a9789932e4e"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_Z</b></td></tr>
<tr class="separator:a62ac5c497b8aeae08e9f3a9789932e4e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0247c14cdec81db76ec294d8dd901715"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0247c14cdec81db76ec294d8dd901715"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_E</b></td></tr>
<tr class="separator:a0247c14cdec81db76ec294d8dd901715"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1d101ab103c56784325777debfacf60"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab1d101ab103c56784325777debfacf60"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>b_differentiation</b></td></tr>
<tr class="separator:ab1d101ab103c56784325777debfacf60"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1dfa9270abb7bbd9d402a48168380eb5"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1dfa9270abb7bbd9d402a48168380eb5"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_pairwiseFixedDifferences</b></td></tr>
<tr class="separator:a1dfa9270abb7bbd9d402a48168380eb5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa121db97c268f61d39220dac03a6350f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa121db97c268f61d39220dac03a6350f"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_pairwiseCommonAlleles</b></td></tr>
<tr class="separator:aa121db97c268f61d39220dac03a6350f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a05a05ea35fafa22c6bde7c45663f3911"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a05a05ea35fafa22c6bde7c45663f3911"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_pairwiseSharedAlleles</b></td></tr>
<tr class="separator:a05a05ea35fafa22c6bde7c45663f3911"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3238e1ff186810718b3fc85288a1a1f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac3238e1ff186810718b3fc85288a1a1f"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_popPolymorphic</b></td></tr>
<tr class="separator:ac3238e1ff186810718b3fc85288a1a1f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1eaa434705bc84a9e0f0c03c5f39378e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1eaa434705bc84a9e0f0c03c5f39378e"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_popSpecific</b></td></tr>
<tr class="separator:a1eaa434705bc84a9e0f0c03c5f39378e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8645b89c29e4503447345b57a2626bc9"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8645b89c29e4503447345b57a2626bc9"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_popSpecificDerived</b></td></tr>
<tr class="separator:a8645b89c29e4503447345b57a2626bc9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0ae7f56d53bd8ba6b02bf1741811230c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0ae7f56d53bd8ba6b02bf1741811230c"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_countFixedDifferences</b></td></tr>
<tr class="separator:a0ae7f56d53bd8ba6b02bf1741811230c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af10bc45cf96ae60db5a53a9feaf3894e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af10bc45cf96ae60db5a53a9feaf3894e"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_countCommonAlleles</b></td></tr>
<tr class="separator:af10bc45cf96ae60db5a53a9feaf3894e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a64c8862b1bc874d2ff0237a21644e95f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a64c8862b1bc874d2ff0237a21644e95f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_countSharedAlleles</b></td></tr>
<tr class="separator:a64c8862b1bc874d2ff0237a21644e95f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf9e060a8485a524f1b70d7312b326b8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adf9e060a8485a524f1b70d7312b326b8"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_countSpecificAlleles</b></td></tr>
<tr class="separator:adf9e060a8485a524f1b70d7312b326b8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4b707e00ca4932ba64a16b9e781cb7f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac4b707e00ca4932ba64a16b9e781cb7f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_countSpecificDerivedAlleles</b></td></tr>
<tr class="separator:ac4b707e00ca4932ba64a16b9e781cb7f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab136497f6b42d3526a95d85f8aaf3b6d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab136497f6b42d3526a95d85f8aaf3b6d"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>b_triConfigurations</b></td></tr>
<tr class="separator:ab136497f6b42d3526a95d85f8aaf3b6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a430220114e843ef51d869a3439523f30"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a430220114e843ef51d869a3439523f30"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_triConfigurations</b></td></tr>
<tr class="separator:a430220114e843ef51d869a3439523f30"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_attribs_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pro_attribs_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Protected Attributes inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a2aa7a2be99f12e471f767cee98d6a1f1 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2aa7a2be99f12e471f767cee98d6a1f1"></a>
<a class="el" href="classegglib_1_1SitePolymorphism.html">SitePolymorphism</a> **&#160;</td><td class="memItemRight" valign="bottom"><b>v_sites</b></td></tr>
<tr class="separator:a2aa7a2be99f12e471f767cee98d6a1f1 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a65260e65d24391c727ff74dafd05411a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a65260e65d24391c727ff74dafd05411a"></a>
bool *&#160;</td><td class="memItemRight" valign="bottom"><b>v_orientables</b></td></tr>
<tr class="separator:a65260e65d24391c727ff74dafd05411a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f24b85598b7bfbd92e59a2aa816468a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0f24b85598b7bfbd92e59a2aa816468a"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_sitePositions</b></td></tr>
<tr class="separator:a0f24b85598b7bfbd92e59a2aa816468a inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acaa4bf5869424fc58e4f454ad32983ef inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acaa4bf5869424fc58e4f454ad32983ef"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_reserved</b></td></tr>
<tr class="separator:acaa4bf5869424fc58e4f454ad32983ef inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a49f7b76263e8d660cad6bab0b1a0ee18 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a49f7b76263e8d660cad6bab0b1a0ee18"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_ns</b></td></tr>
<tr class="separator:a49f7b76263e8d660cad6bab0b1a0ee18 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1d5cf6776213289afae59d1d5013ef3 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab1d5cf6776213289afae59d1d5013ef3"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_S</b></td></tr>
<tr class="separator:ab1d5cf6776213289afae59d1d5013ef3 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a951aa44c6e7034205b92087a5b0ad1be inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a951aa44c6e7034205b92087a5b0ad1be"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_So</b></td></tr>
<tr class="separator:a951aa44c6e7034205b92087a5b0ad1be inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ad85c1898fbeba3f7a795a5d8901812 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6ad85c1898fbeba3f7a795a5d8901812"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_eta</b></td></tr>
<tr class="separator:a6ad85c1898fbeba3f7a795a5d8901812 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa3ed42221d826288582cd5d4ffe561b2 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa3ed42221d826288582cd5d4ffe561b2"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_nseff</b></td></tr>
<tr class="separator:aa3ed42221d826288582cd5d4ffe561b2 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb38ffa03ab28fcdf786dbad7cf3c800 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aeb38ffa03ab28fcdf786dbad7cf3c800"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_lseff</b></td></tr>
<tr class="separator:aeb38ffa03ab28fcdf786dbad7cf3c800 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f748a5c62571140a7a3df61711190f4 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7f748a5c62571140a7a3df61711190f4"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>v_nseffo</b></td></tr>
<tr class="separator:a7f748a5c62571140a7a3df61711190f4 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c42f196ad1fb4cdf0031bb4acc53d6f inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3c42f196ad1fb4cdf0031bb4acc53d6f"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_lseffo</b></td></tr>
<tr class="separator:a3c42f196ad1fb4cdf0031bb4acc53d6f inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3f8ba820cd11d01df127e058ab4710b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac3f8ba820cd11d01df127e058ab4710b"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>v_npop</b></td></tr>
<tr class="separator:ac3f8ba820cd11d01df127e058ab4710b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8b67942007a1d99348372b3fd25e65b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8b67942007a1d99348372b3fd25e65b"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><b>v_popLabel</b></td></tr>
<tr class="separator:ac8b67942007a1d99348372b3fd25e65b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac93cdae4eb273f9e3641d94eb90d93c9 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac93cdae4eb273f9e3641d94eb90d93c9"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>p_allowMultipleMutations</b></td></tr>
<tr class="separator:ac93cdae4eb273f9e3641d94eb90d93c9 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a971874b4d9cb1c5737ec7e1d73c4a6f5 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a971874b4d9cb1c5737ec7e1d73c4a6f5"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>p_minimumExploitableData</b></td></tr>
<tr class="separator:a971874b4d9cb1c5737ec7e1d73c4a6f5 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb27bce76989a6521804ba5f618efe87 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb27bce76989a6521804ba5f618efe87"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><b>p_characterMapping</b></td></tr>
<tr class="separator:afb27bce76989a6521804ba5f618efe87 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a12a9ec49e72be17b7c4993df7ff20903 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a12a9ec49e72be17b7c4993df7ff20903"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>p_pos_sep_mapping</b></td></tr>
<tr class="separator:a12a9ec49e72be17b7c4993df7ff20903 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58c7dc12609f51192588cbc5b08c1f5b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a58c7dc12609f51192588cbc5b08c1f5b"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>p_useZeroAsAncestral</b></td></tr>
<tr class="separator:a58c7dc12609f51192588cbc5b08c1f5b inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acdeb3544a8b488558119e39ed5b27035 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="acdeb3544a8b488558119e39ed5b27035"></a>
unsigned int&#160;</td><td class="memItemRight" valign="bottom"><b>p_ignoreFrequency</b></td></tr>
<tr class="separator:acdeb3544a8b488558119e39ed5b27035 inherit pro_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="inherited"></a>
Additional Inherited Members</h2></td></tr>
<tr class="inherit_header pub_static_attribs_classegglib_1_1BaseDiversity"><td colspan="2" onclick="javascript:toggleInherit('pub_static_attribs_classegglib_1_1BaseDiversity')"><img src="closed.png" alt="-"/>&#160;Static Public Attributes inherited from <a class="el" href="classegglib_1_1BaseDiversity.html">BaseDiversity</a></td></tr>
<tr class="memitem:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a> = &quot;ACGT MRWSYKBDHVN?-&quot;</td></tr>
<tr class="memdesc:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for DNA data.  <a href="#a1e968823aeb9f9c279596cc3c2b93431">More...</a><br /></td></tr>
<tr class="separator:a1e968823aeb9f9c279596cc3c2b93431 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#a481307ebf500b6b714e767694019be94">rnaMapping</a> = &quot;ACGU MRWSYKBDHVN?-&quot;</td></tr>
<tr class="memdesc:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for RNA data.  <a href="#a481307ebf500b6b714e767694019be94">More...</a><br /></td></tr>
<tr class="separator:a481307ebf500b6b714e767694019be94 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memItemLeft" align="right" valign="top">static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classegglib_1_1BaseDiversity.html#aa93366ce287a12c730120fc0e1cd2875">aaMapping</a> = &quot;GALMFWKQESPVICYHRNDT X-&quot;</td></tr>
<tr class="memdesc:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predefined mapping string for amino acid data.  <a href="#aa93366ce287a12c730120fc0e1cd2875">More...</a><br /></td></tr>
<tr class="separator:aa93366ce287a12c730120fc0e1cd2875 inherit pub_static_attribs_classegglib_1_1BaseDiversity"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Performs analyzes of population genetics. </p>
<p>This class computes several summary statistics based on nucleotide analysis. Note that it is possible to use the same object to analyze different data set. Calling the <a class="el" href="classegglib_1_1NucleotideDiversity.html#a0f7781582853b62c254596290a0c0af1" title="Identifies polymorphic sites and computes basis statistics. ">load()</a> method erases all data preivously computed (if any). Calling the <a class="el" href="classegglib_1_1NucleotideDiversity.html#a0f7781582853b62c254596290a0c0af1" title="Identifies polymorphic sites and computes basis statistics. ">load()</a> method is absolutely required to compute any statistics. Some statistics are not computed by default, but are if the corresponding accessor is used (only <a class="el" href="classegglib_1_1NucleotideDiversity.html#a0f7781582853b62c254596290a0c0af1" title="Identifies polymorphic sites and computes basis statistics. ">load()</a> is required).</p>
<p>Note that "unsecure" accessors don't perform out-of-bound checks.</p>
<p>S is the number of varying sites (only in sites that were not rejected).</p>
<p>eta is the minimum number of mutations, that is the sum of the number of alleles minus 1 for each varying site. eta = S if all sites have no variant or 2 alleles. eta is computed independently of the option multiple and IS NOT computed over lseff sites.</p>
<p>Pi is the average number of pairwise differences between sequences (expressed here per site) or (as computed here) the mean per site (unbiased) heterozygosity. Pi is zero if no polymorphic sites.</p>
<p>D is the Tajima's test of neutrality Ref. Tajima F.: Statistical method for testing the neutral mutation hypothesis by DNA polymorphism. Genetics 1989, 123:585-595. It is arbitrary set to 0 if no polymorphic sites.</p>
<p>tW: thetaW: estimator of theta based on polymorphic sites (ref. e.g. Watterson 1975 Theor. Pop. Biol.). Both D and thetaW are computed assuming that rounded nseff samples have been sampled. The variance of D is computed using rounded nseff instead of ns.</p>
<p>H is the Fay and Wu's test of neutrality. Z is the standardized version and E a similar test. Ref. Fay J. C., Wu C.-I.: Hitchhiking under positive Darwinian selection. Genetics 2000, 155:1405-1413. and Zeng K., Fu Y. X., Shi S., Wu C.-I.: Statistical tests for detecting positive selection by utilizing high-frequency variants. Genetics 2006, 174:1431-9. Both are arbitrary set to 0 if no polymorphic or orientable sites.</p>
<p>tH and tL: theta H: estimators of theta based on derived polymorphic sites (ref in Fay and Wu and Zeng al.). The variance of H/Z are computed assuming that rounded nseff samples have been sampled. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a4b81fe7e6f785b5413d9bfd192fa39e3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Builds an object. </p>

</div>
</div>
<a class="anchor" id="aadecb2de0968a245ae1b36543bdec89b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">~<a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Destroys an object. </p>

</div>
</div>
<a class="anchor" id="abd422da68dace0766ec6df73032b0974"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> &amp;&#160;</td>
          <td class="paramname"><em>source</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be copied. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a0f7781582853b62c254596290a0c0af1"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void load </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classegglib_1_1CharMatrix.html">CharMatrix</a> &amp;&#160;</td>
          <td class="paramname"><em>data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allowMultipleMutations</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>minimumExploitableData</em> = <code>1.</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>ignoreFrequency</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>characterMapping</em> = <code><a class="el" href="classegglib_1_1BaseDiversity.html#a1e968823aeb9f9c279596cc3c2b93431">dnaMapping</a></code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>useZeroAsAncestral</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Identifies polymorphic sites and computes basis statistics. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">data</td><td>an alignment object (subclass of <a class="el" href="classegglib_1_1CharMatrix.html" title="Interface for classes usable as a square matrix of characters. ">CharMatrix</a>). The presence of outgroup or of different populations will be detected based on the populationLabel members of the passed object. The populationLabel 999 will be interpreted as outgroups. If several outgroups are passed, sites were the outgroups are not consistent will be treated as "non-
orientable".</td></tr>
    <tr><td class="paramname">allowMultipleMutations</td><td>if true, sites with more than two alleles will not be ignored. The sum of the frequencies of all alleles not matching the outgroup will treated as the derived allele frequency (for orientable sites).</td></tr>
    <tr><td class="paramname">minimumExploitableData</td><td>sites where the non-missing data (as defined by characterMapping) are at a frequency larger than this value will be removed from the analysis. Use 1. to take only 'complete' sites into account and 0. to use all sites. (The outgroup is not considered in this computation.)</td></tr>
    <tr><td class="paramname">ignoreFrequency</td><td>removes sites that are polymorph because of an allele at absolute frequency smaller than or equal to this value. If ignoreFrequency=1, no sites are removed, if ignoreFrequency=0, singleton sites are ignored. Such sites are completely removed from the analysis (not counted in lseff). Note that if more than one mutation is allowed, the site is removed only if all the alleles but one are smaller than or equal to this value. For example, an alignment column AAAAAAGAAT is ignored with an ignoreFrequency of 1, but AAAAAAGGAT is conserved (including the third allele T which is a singleton).</td></tr>
    <tr><td class="paramname">characterMapping</td><td>a string giving the list of characters that should be considered as valid data. If a space is present in the string, the characters left of the space will be treated as valid data and the characters right of the space will be treated as missing data, that is tolerated but ignored. All characters not in the string will cause an <a class="el" href="classegglib_1_1EggInvalidCharacterError.html" title="Exception type for invalid character. ">EggInvalidCharacterError</a> to be raised.</td></tr>
    <tr><td class="paramname">useZeroAsAncestral</td><td>if true, all outgroups (if present) will be ignored and the character "0" will be considered as ancestral for all sites, whatever the character mapping. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a426e6e4a99cfc9c89887f24a28af9d3e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a>&amp; operator= </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classegglib_1_1NucleotideDiversity.html">NucleotideDiversity</a> &amp;&#160;</td>
          <td class="paramname"><em>source</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>This class cannot be copied. </p>

</div>
</div>
<a class="anchor" id="adc71bbf66c1d95cc029c03e811457825"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; unsigned int &gt; singleton_positions </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Builds and returns the vector of positions of all singleton sites. </p>
<p>A site singleton when it is polymorphic according to parameter of the diversity analysis, when it has exactly two alleles and one of them is at absolute frequency 1 (one copy) disregarding the outgroup. </p>

</div>
</div>
<a class="anchor" id="a011ed8f7cd114c5665a3444ebd68a24f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int triConfiguration </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Number falling into one of the possible site configurations. </p>
<p>The statistics are limited to three populations. Assuming an unrooted A/G polymorphism (A and G can be substitued), the site configurations are:</p><ul>
<li>0: A&amp;G A A specific 1</li>
<li>1: A&amp;G A G specific 1 + fixed 2-3</li>
<li>2: A A&amp;G A specific 2</li>
<li>3: A A&amp;G G specific 2 + fixed 1-3</li>
<li>4: A A A&amp;G specific 3</li>
<li>5: A G A&amp;G specific 3 + fixed 1-2</li>
<li>6: A&amp;G A&amp;G A shared 1-2</li>
<li>7: A&amp;G A A&amp;G shared 1-3</li>
<li>8: A A&amp;G A&amp;G shared 2-3</li>
<li>9: A&amp;G A&amp;G A&amp;G shared 1-2-3</li>
<li>10: A G G fixed 1</li>
<li>11: A G A fixed 2</li>
<li>12: A A G fixed 3</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">index</td><td>must be an index from 0 to 12. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/<a class="el" href="NucleotideDiversity_8hpp_source.html">NucleotideDiversity.hpp</a></li>
<li>/home/stephane/Documents/devel/egglib2/egglib-cpp/NucleotideDiversity.cpp</li>
</ul>
</div><!-- contents -->
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head></head><body><hr size="1">
<address style="text-align: right;"><small>Hosted by&nbsp;</address>
<a href="http://sourceforge.net/projects/egglib"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=232965&amp;type=8" width="80" height="15" alt="Get EggLib at SourceForge.net. Fast, secure and Free Open Source software downloads" / ALIGN=RIGHT BORDER=0></a>
</body></html>
