// wollok-ts/dist/printer/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultToEmpty = exports.stringify = exports.encloseIndented = exports.enclosedList = exports.listed = exports.body = exports.WS = exports.setEnclosers = exports.listEnclosers = void 0;
const prettier_printer_1 = require("prettier-printer");
exports.listEnclosers = prettier_printer_1.brackets;
exports.setEnclosers = ['#{', '}'];
exports.WS = ' ';
const body = (nest) => (content) => (0, exports.encloseIndented)(prettier_printer_1.braces, content, nest);
exports.body = body;
const listed = (contents, separator = ',') => (0, prettier_printer_1.intersperse)([separator, prettier_printer_1.softLine], contents);
exports.listed = listed;
const enclosedList = (nest) => (enclosers, content, forceSpread = false, separator = ',') => {
    if (content.length === 0)
        return (0, prettier_printer_1.enclose)(enclosers, '');
    const narrowFormat = (0, exports.encloseIndented)(['', ''], (0, prettier_printer_1.intersperse)([separator, prettier_printer_1.lineBreak], content), nest);
    return (0, prettier_printer_1.enclose)(enclosers, forceSpread ? narrowFormat : (0, prettier_printer_1.choice)((0, prettier_printer_1.intersperse)([separator, exports.WS], content), narrowFormat));
};
exports.enclosedList = enclosedList;
const encloseIndented = (enclosers, content, nest) => (0, prettier_printer_1.enclose)(enclosers, (0, prettier_printer_1.append)(prettier_printer_1.lineBreak, nest([prettier_printer_1.lineBreak, content])));
exports.encloseIndented = encloseIndented;
exports.stringify = (0, prettier_printer_1.enclose)(prettier_printer_1.dquotes);
const defaultToEmpty = (condition, doc) => condition ? doc : [];
exports.defaultToEmpty = defaultToEmpty;
